# Τεχνητή Νοημοσύνη: Εργαστηριακή Άσκηση 2
---

Ο στόχος της εργασίας είναι η κατασκευή ενός συστήματος προτάσεων (Recommendation System) για ταινίες. Οι προτάσεις αυτές θα πηγάζουν τόσο από τα χαρακτηριστικά της ταινίας όσο και από ορισμένες αξιολογήσεις του κάθε χρήστη.




# **Μέρος 1: Μελέτη των Metadata, Δημιουργία κόσμου και των βασικών queries.**


##Ερώτημα  1α

Μελετάμε την δομή και τα χαρακτηριστικά του αρχείου movies_metadata.csv. Βλέπουμε πως το αρχείο περιλαμβάνει μία λίστα από ταινίες και για καθεμία από αυτές ένα σύνολο χαρακτηρηστικών, όπως ο τίτλος, το είδος, ο σκηνοθέτης κ.α.  Με βάση αυτό, κατασκευάζουμε μία βάση γνώσης στην Prolog:

```
budget(Movie,Budget).
genre(Movie,Genre).
homepage(Movie,Homepage).
id(Movie,ID).
plot_keyword(Movie,Plot_keyword).
language(Movie,Language).
original_title(Movie,Title).
overview(Movie,Overview).
popularity(Movie,Popularity).
production_company(Movie,Company,Id).
production_country(Movie,Iso_3166_1,Name).
release_date(Movie,Release_date).
gross(Movie,Gross).
duration(Movie,Duration).
spoken_language(Movie,Iso_639_1,Name).
status(Movie,Status).
tagline(Movie,Tagline).
movie_title(Movie,Title).
vote_average(Movie,Vote_average).
num_voted_users(Movie,Num_voted_users).
title_year(Movie,Title_year).
country(Movie,Country).
director_name(Movie,Director_name).
actor_1_name(Movie,Actor_1_name).
actor_2_name(Movie,Actor_2_name).
actor_3_name(Movie,Actor_3_name).
```

##Ερώτημα 1β

Έχοντας λοιπόν δημιουργήσει τον κόσμο του προβλήματος δημιουργούμε ένα σύνολο από απλούς κανόνες οι οποίοι θα βρίσκουν όλες τις ταινίες με:
1.	Κοινό θέμα (κάποιες λέξεις σχετικά με το genre κοινές)
2.	Αρκετά κοινό θέμα (κάποιες λιγότερες λέξεις σχετικά με το genre κοινές π.χ. 3)
3.	Σχετικά κοινό θέμα (λίγες λέξεις σχετικές με το genre κοινές π.χ. 1)
4.	Κοινός σκηνοθέτης
5.	Ακριβώς ίδια πλοκή (κάποιες λέξεις κλειδιά της πλοκής κοινές)
6.	Σχετικά ίδια πλοκή (κάποιες λιγότερες λέξεις κλειδιά κοινές)
7.	Ίδιους τους βασικούς ηθοποιούς (και τους 3)
8.	Αρκετά ίδιους βασικούς ηθοποιούς (ορισμένους βασικούς ηθοποιούς κοινούς π.χ. 2)
9.	Σχετικά ίδιους ηθοποιούς (π.χ. 1 από τους 3)
10.	Ίδια γλώσσα
11.	Είναι έγχρωμες ή ασπρόμαυρες
12. Κοινό studio παραγωγής
13. Κοινή χώρα παραγωγής 
14. Ίδια δεκαετία 

Προσθέτουμε επιπλέον ορισμένους κανόνες, σε περιπτώσεις ομοιότητας η οποία λαμβάνει κάποια κλίμακα (π.χ. ένα, δύο, τρία κλπ κοινά genres), ώστε να έχουμε πιο μεγάλη πληρότητα, καθώς και ορισμένους κανόνες που σχετίζονται με άλλα δοσμένα χαρακτηρηστικά, π.χ. vote_average, οι οποίοι δεν ζητώνται παραπάνω:

> Κανόνας 1

```
five_common_genre(Movie_1,Movie_2):-
  genre(Movie_1,X),
  genre(Movie_2,X),
  genre(Movie_1,Y),
  genre(Movie_2,Y),
  genre(Movie_1,Z),
  genre(Movie_2,Z),
  genre(Movie_1,W),
  genre(Movie_2,W),
  genre(Movie_1,R),
  genre(Movie_2,R),
  Movie_1 \= Movie_2,
  X \= Y,
  X \= Z,
  X \= W,
  X \= R,
  Y \= Z,
  Y \= W,
  Y \= R,
  Z \= W,
  Z \= R,
  W \= R.
```

> Κανόνας 2

```
three_common_genre(Movie_1,Movie_2):-
  genre(Movie_1,X),
  genre(Movie_2,X),
  genre(Movie_1,Y),
  genre(Movie_2,Y),
  genre(Movie_1,Z),
  genre(Movie_2,Z),
  Movie_1 \= Movie_2,
  X \= Y,
  X \= Z,
  Y \= Z.
```

> Κανόνας 3

```
one_common_genre(Movie_1,Movie_2):-
  genre(Movie_1,X),
  genre(Movie_2,X),
  Movie_1 \= Movie_2.
```

**Προσθέτουμε επίσης:** 
```
two_common_genre(Movie_1,Movie_2):-
  genre(Movie_1,X),
  genre(Movie_2,X),
  genre(Movie_1,Y),
  genre(Movie_2,Y),
  Movie_1 \= Movie_2,
  X \= Y.

```
```
four_common_genre(Movie_1,Movie_2):-
  genre(Movie_1,X),
  genre(Movie_2,X),
  genre(Movie_1,Y),
  genre(Movie_2,Y),
  genre(Movie_1,Z),
  genre(Movie_2,Z),
  genre(Movie_1,W),
  genre(Movie_2,W),
  Movie_1 \= Movie_2,
  X \= Y,
  X \= Z,
  X \= W,
  Y \= Z,
  Y \= W,
  Z \= W.

```

> Κανόνας 4

```
common_director(Movie_1,Movie_2):-
  director_name(Movie_1,X),
  director_name(Movie_2,X),
  Movie_1 \= Movie_2.
```
> Κανόνας 5

```
five_same_plot(Movie_1,Movie_2):-
  plot_keyword(Movie_1,X),
  plot_keyword(Movie_2,X),
  plot_keyword(Movie_1,Y),
  plot_keyword(Movie_2,Y),
  plot_keyword(Movie_1,Z),
  plot_keyword(Movie_2,Z),
  plot_keyword(Movie_1,W),
  plot_keyword(Movie_2,W),
  plot_keyword(Movie_1,R),
  plot_keyword(Movie_2,R),
  Movie_1 \= Movie_2,
  X \= Y,
  X \= Z,
  X \= W,
  X \= R,
  Y \= Z,
  Y \= W,
  Y \= R,
  Z \= W,
  Z \= R,
  W \= R.
```
> Κανόνας 6

```
three_same_plot(Movie_1,Movie_2):-
  plot_keyword(Movie_1,X),
  plot_keyword(Movie_2,X),
  plot_keyword(Movie_1,Y),
  plot_keyword(Movie_2,Y),
  plot_keyword(Movie_1,Z),
  plot_keyword(Movie_2,Z),
  Movie_1 \= Movie_2,
  X \= Y,
  X \= Z,
  Y \= Z.
```
**Προσθέτουμε επίσης:**

```
one_same_plot(Movie_1,Movie_2):-
  plot_keyword(Movie_1,X),
  plot_keyword(Movie_2,X),
  Movie_1 \= Movie_2.
```
```
two_same_plot(Movie_1,Movie_2):-
  plot_keyword(Movie_1,X),
  plot_keyword(Movie_2,X),
  plot_keyword(Movie_1,Y),
  plot_keyword(Movie_2,Y),
  Movie_1 \= Movie_2,
  X \= Y.
```
```
four_same_plot(Movie_1,Movie_2):-
  plot_keyword(Movie_1,X),
  plot_keyword(Movie_2,X),
  plot_keyword(Movie_1,Y),
  plot_keyword(Movie_2,Y),
  plot_keyword(Movie_1,Z),
  plot_keyword(Movie_2,Z),
  plot_keyword(Movie_1,W),
  plot_keyword(Movie_2,W),
  Movie_1 \= Movie_2,
  X \= Y,
  X \= Z,
  X \= W,
  Y \= Z,
  Y \= W,
  Z \= W.
```
> Κανόνας 7

```
same_actors(Movie_1,Movie_2):-
  actor_1_name(Movie_1,X),actor_1_name(Movie_2,X),actor_2_name(Movie_1,Y),actor_2_name(Movie_2,Y),actor_3_name(Movie_1,Z),actor_3_name(Movie_2,Z),Movie_1 \= Movie_2;
  actor_1_name(Movie_1,X),actor_2_name(Movie_2,X),actor_2_name(Movie_1,Y),actor_1_name(Movie_2,Y),actor_3_name(Movie_1,Z),actor_3_name(Movie_2,Z),Movie_1 \= Movie_2;
  actor_1_name(Movie_1,X),actor_3_name(Movie_2,X),actor_2_name(Movie_1,Y),actor_2_name(Movie_2,Y),actor_3_name(Movie_1,Z),actor_1_name(Movie_2,Z),Movie_1 \= Movie_2;
  actor_1_name(Movie_1,X),actor_1_name(Movie_2,X),actor_2_name(Movie_1,Y),actor_3_name(Movie_2,Y),actor_3_name(Movie_1,Z),actor_2_name(Movie_2,Z),Movie_1 \= Movie_2;
  actor_1_name(Movie_1,X),actor_2_name(Movie_2,X),actor_2_name(Movie_1,Y),actor_3_name(Movie_2,Y),actor_3_name(Movie_1,Z),actor_1_name(Movie_2,Z),Movie_1 \= Movie_2;
  actor_1_name(Movie_1,X),actor_3_name(Movie_2,X),actor_2_name(Movie_1,Y),actor_1_name(Movie_2,Y),actor_3_name(Movie_1,Z),actor_2_name(Movie_2,Z),Movie_1 \= Movie_2.
```
> Κανόνας 8

```
some_same_actors(Movie_1,Movie_2):-
  actor_1_name(Movie_1,X),actor_1_name(Movie_2,X),actor_2_name(Movie_1,Y),actor_2_name(Movie_2,Y),Movie_1 \= Movie_2;
  actor_1_name(Movie_1,X),actor_2_name(Movie_2,X),actor_2_name(Movie_1,Y),actor_1_name(Movie_2,Y),Movie_1 \= Movie_2;
  actor_1_name(Movie_1,X),actor_2_name(Movie_2,X),actor_2_name(Movie_1,Y),actor_3_name(Movie_2,Y),Movie_1 \= Movie_2;
  actor_1_name(Movie_1,X),actor_3_name(Movie_2,X),actor_3_name(Movie_1,Y),actor_2_name(Movie_2,Y),Movie_1 \= Movie_2;
  actor_1_name(Movie_1,X),actor_3_name(Movie_2,X),actor_2_name(Movie_1,Y),actor_1_name(Movie_2,Y),Movie_1 \= Movie_2;
  actor_1_name(Movie_1,X),actor_1_name(Movie_2,X),actor_2_name(Movie_1,Y),actor_3_name(Movie_2,Y),Movie_1 \= Movie_2;
  actor_2_name(Movie_1,X),actor_1_name(Movie_2,X),actor_3_name(Movie_1,Y),actor_2_name(Movie_2,Y),Movie_1 \= Movie_2;
  actor_2_name(Movie_1,X),actor_2_name(Movie_2,X),actor_3_name(Movie_1,Y),actor_1_name(Movie_2,Y),Movie_1 \= Movie_2;
  actor_2_name(Movie_1,X),actor_2_name(Movie_2,X),actor_3_name(Movie_1,Y),actor_3_name(Movie_2,Y),Movie_1 \= Movie_2;
  actor_2_name(Movie_1,X),actor_2_name(Movie_2,X),actor_1_name(Movie_1,Y),actor_3_name(Movie_2,Y),Movie_1 \= Movie_2;
  actor_2_name(Movie_1,X),actor_3_name(Movie_2,X),actor_3_name(Movie_1,Y),actor_2_name(Movie_2,Y),Movie_1 \= Movie_2;
  actor_2_name(Movie_1,X),actor_3_name(Movie_2,X),actor_3_name(Movie_1,Y),actor_1_name(Movie_2,Y),Movie_1 \= Movie_2;
  actor_2_name(Movie_1,X),actor_1_name(Movie_2,X),actor_3_name(Movie_1,Y),actor_3_name(Movie_2,Y),Movie_1 \= Movie_2;
  actor_3_name(Movie_1,X),actor_1_name(Movie_2,X),actor_1_name(Movie_1,Y),actor_2_name(Movie_2,Y),Movie_1 \= Movie_2;
  actor_3_name(Movie_1,X),actor_2_name(Movie_2,X),actor_1_name(Movie_1,Y),actor_1_name(Movie_2,Y),Movie_1 \= Movie_2;
  actor_3_name(Movie_1,X),actor_3_name(Movie_2,X),actor_1_name(Movie_1,Y),actor_2_name(Movie_2,Y),Movie_1 \= Movie_2;
  actor_3_name(Movie_1,X),actor_3_name(Movie_2,X),actor_1_name(Movie_1,Y),actor_1_name(Movie_2,Y),Movie_1 \= Movie_2;
  actor_3_name(Movie_1,X),actor_1_name(Movie_2,X),actor_1_name(Movie_1,Y),actor_3_name(Movie_2,Y),Movie_1 \= Movie_2.
```
> Κανόνας 9

```
one_same_actor(Movie_1,Movie_2):-
  actor_1_name(Movie_1,X),actor_1_name(Movie_2,X),Movie_1 \= Movie_2;
  actor_1_name(Movie_1,X),actor_2_name(Movie_2,X),Movie_1 \= Movie_2;
  actor_1_name(Movie_1,X),actor_3_name(Movie_2,X),Movie_1 \= Movie_2;
  actor_2_name(Movie_1,X),actor_1_name(Movie_2,X),Movie_1 \= Movie_2;
  actor_2_name(Movie_1,X),actor_2_name(Movie_2,X),Movie_1 \= Movie_2;
  actor_2_name(Movie_1,X),actor_3_name(Movie_2,X),Movie_1 \= Movie_2;
  actor_3_name(Movie_1,X),actor_3_name(Movie_2,X),Movie_1 \= Movie_2;
  actor_3_name(Movie_1,X),actor_2_name(Movie_2,X),Movie_1 \= Movie_2;
  actor_3_name(Movie_1,X),actor_3_name(Movie_2,X),Movie_1 \= Movie_2.
```
> Κανόνας 10

```
same_language(Movie_1,Movie_2):-
  language(Movie_1,X),
  language(Movie_2,X),
  Movie_1 \= Movie_2.
```

> Κανόνας 11  (black_and_white)

Στον κανόνα αυτό χρησιμοποιούμε ορισμένους βοηθητικούς κανόνες. Ο κανόνας find_black βρίσκει όλα τα ζεύγη ταινιών που είναι ασπρόμαυρες, ο κανόνας is_black_and_white ελέγχει αν μία ταινία είναι ασπρόμαυρη, και ο κανόνας find_color βρίσκει για μία δοσμένη έγχρωμη ταινία όλες τις υπόλοιπες εκτός αυτής οι οποίες είναι έγχρωμες. Τέλος, ο κανόνας black_and_white βρίσκει τις ταινίες που είναι όμοιες, δηλαδή είναι και οι δύο έγχρωμες ή και οι δύο ασπρόμαυρες.

```
find_black(Movie_1,Movie_2):-
  plot_keyword(Movie_1,'black and white'),
  plot_keyword(Movie_2,'black and white'), 
  Movie_1 \= Movie_2.
 
is_black_and_white(Movie):-
  plot_keyword(Movie,'black and white').


find_color(Movie_1,Movie_2):- 
  plot_keyword(Movie_2, Y), ( is_black_and_white(Movie_2) -> false ;
  Movie_1 \= Movie_2
  ).
  
black_and_white(Movie_1,Movie_2):- (
    is_black_and_white(Movie_1) -> find_black(Movie_1,Movie_2);
    find_color(Movie_1,Movie_2)
    ).
```

> Κανόνας 12

```
one_common_production_studio(Movie_1,Movie_2):-
  production_company(Movie_1,X,Y),
  production_company(Movie_2,X,Y),
  Movie_1 \= Movie_2.
```

```
two_common_production_studio(Movie_1,Movie_2):-
  production_company(Movie_1,X,Y),
  production_company(Movie_2,X,Y),
  production_company(Movie_1,Z,W),
  production_company(Movie_2,Z,W),
  X \= Z,
  Y \= W,
  Movie_1 \= Movie_2.
```
```
three_common_production_studio(Movie_1,Movie_2):-
  production_company(Movie_1,X,Y),
  production_company(Movie_2,X,Y),
  production_company(Movie_1,Z,W),
  production_company(Movie_2,Z,W),
  production_company(Movie_1,Q,R),
  production_company(Movie_2,Q,R),
  X \= Z,
  Z \= Q,
  X \= Q,
  Y \= W,
  Y \= R,
  W \= R,
  Movie_1 \= Movie_2.
```
> Κανόνας 13

```
one_common_production_country(Movie_1,Movie_2):-
  production_country(Movie_1,X,Y),
  production_country(Movie_2,X,Y),
  Movie_1 \= Movie_2.
```
```
two_common_production_country(Movie_1,Movie_2):-
  production_country(Movie_1,X,Y),
  production_country(Movie_2,X,Y),
  production_country(Movie_1,Z,W),
  production_country(Movie_2,Z,W),
  X \= Z,
  Y \= W,
  Movie_1 \= Movie_2.
```
```
three_common_production_country(Movie_1,Movie_2):-
  production_country(Movie_1,X,Y),
  production_country(Movie_2,X,Y),
  production_country(Movie_1,Z,W),
  production_country(Movie_2,Z,W),
  production_country(Movie_1,Q,R),
  production_country(Movie_2,Q,R),
  X \= Z,
  X \= Q,
  Z \= Q,
  Y \= W,
  Y \= R,
  W \= R,
  Movie_1 \= Movie_2.
```
> Κανόνας 14

```
common_decade(Movie_1,Movie_2):-
  title_year(Movie_1,X),
  title_year(Movie_2,Y),
  Movie_1 \= Movie_2,
  X1 is round(X),
  Y1 is round(Y),
  Z is div(X1,10),
  W is div(Y1,10),
  Z =:= W.
```

###**Προθέτουμε τους παρακάτω επιπλέον κανόνες :**

> Κοινό budget

```
very_common_budget(Movie_1,Movie_2):-
  budget(Movie_1,Budget_1),
  budget(Movie_2,Budget_2),
  Movie_1 \= Movie_2,
  Dif is Budget_1-Budget_2,
  Diff_abs is abs(Dif),
  Diff_abs =< 1000000.
```
```
partly_common_budget(Movie_1,Movie_2):-
  budget(Movie_1,Budget_1),
  budget(Movie_2,Budget_2),
  Movie_1 \= Movie_2,
  Dif is Budget_1-Budget_2,
  Diff_abs is abs(Dif),
  Diff_abs =< 50000000.
```

> Κοινή δημοφιλία

```
very_common_popularity(Movie_1,Movie_2):-
  popularity(Movie_1,X),
  popularity(Movie_2,Y),
  Movie_1 \= Movie_2,
  Dif is X-Y,
  Diff_abs is abs(Dif),
  Diff_abs =< 10.
```
```
partly_common_popularity(Movie_1,Movie_2):-
  popularity(Movie_1,X),
  popularity(Movie_2,Y),
  Movie_1 \= Movie_2,
  Dif is X-Y,
  Diff_abs is abs(Dif),
  Diff_abs =< 30.
```

> Κοινά μικτά έσοδα

```
very_common_gross(Movie_1,Movie_2):-
  gross(Movie_1,Gross_1),
  gross(Movie_2,Gross_2),
  Movie_1 \= Movie_2,
  Dif is Gross_1-Gross_2,
  Diff_abs is abs(Dif),
  Diff_abs =< 1000000.
```
```
partly_common_gross(Movie_1,Movie_2):-
  gross(Movie_1,Gross_1),
  gross(Movie_2,Gross_2),
  Movie_1 \= Movie_2,
  Dif is Gross_1-Gross_2,
  Diff_abs is abs(Dif),
  Diff_abs =< 50000000.
```

> Κοινή διάρκεια

```
common_duration(Movie_1,Movie_2):-
  duration(Movie_1,X),
  duration(Movie_2,Y),
  Movie_1 \= Movie_2,
  Dif is X-Y,
  Diff_abs is abs(Dif),
  Diff_abs =< 30.
```

> Κοινές ομιλούμενες γλώσσες

```
five_common_spoken_language(Movie_1,Movie_2):-
  spoken_language(Movie_1,X1,X2),
  spoken_language(Movie_2,X1,X2),
  spoken_language(Movie_1,Y1,Y2),
  spoken_language(Movie_2,Y1,Y2),
  spoken_language(Movie_1,Z1,Z2),
  spoken_language(Movie_2,Z1,Z2),
  spoken_language(Movie_1,W1,W2),
  spoken_language(Movie_2,W1,W2),
  spoken_language(Movie_1,R1,R2),
  spoken_language(Movie_2,R1,R2),
  Movie_1 \= Movie_2,
  X1 \= Y1,
  X2 \= Y2,
  X1 \= Z1,
  X2 \= Z2,
  X1 \= W1,
  X2 \= W2,
  X1 \= R1,
  X2 \= R2,
  Y1 \= Z1,
  Y2 \= Z2,
  Y1 \= W1,
  Y2 \= W2,
  Y1 \= R1,
  Y2 \= R2,
  Z1 \= W1,
  Z2 \= W2,
  Z1 \= R1,
  Z2 \= R2,
  R1 \= W1,
  R2 \= W2.
```
```
four_common_spoken_language(Movie_1,Movie_2):-
  spoken_language(Movie_1,X1,X2),
  spoken_language(Movie_2,X1,X2),
  spoken_language(Movie_1,Y1,Y2),
  spoken_language(Movie_2,Y1,Y2),
  spoken_language(Movie_1,Z1,Z2),
  spoken_language(Movie_2,Z1,Z2),
  spoken_language(Movie_1,W1,W2),
  spoken_language(Movie_2,W1,W2),
  Movie_1 \= Movie_2,
  X1 \= Y1,
  X2 \= Y2,
  X1 \= Z1,
  X2 \= Z2,
  X1 \= W1,
  X2 \= W2,
  Y1 \= Z1,
  Y2 \= Z2,
  Y1 \= W1,
  Y2 \= W2,
  Z1 \= W1,
  Z2 \= W2.
```
```
three_common_spoken_language(Movie_1,Movie_2):-
  spoken_language(Movie_1,X1,X2),
  spoken_language(Movie_2,X1,X2),
  spoken_language(Movie_1,Y1,Y2),
  spoken_language(Movie_2,Y1,Y2),
  spoken_language(Movie_1,Z1,Z2),
  spoken_language(Movie_2,Z1,Z2),
  Movie_1 \= Movie_2,
  X1 \= Y1,
  X2 \= Y2,
  X1 \= Z1,
  X2 \= Z2,
  Y1 \= Z1,
  Y2 \= Z2.
```
```
two_common_spoken_language(Movie_1,Movie_2):-
  spoken_language(Movie_1,X1,X2),
  spoken_language(Movie_2,X1,X2),
  spoken_language(Movie_1,Y1,Y2),
  spoken_language(Movie_2,Y1,Y2),
  Movie_1 \= Movie_2,
  X1 \= Y1,
  X2 \= Y2.
```
```
one_common_spoken_language(Movie_1,Movie_2):-
  spoken_language(Movie_1,X1,X2),
  spoken_language(Movie_2,X1,X2),
  Movie_1 \= Movie_2.
```

> Κοινή μέση βαθμολογία

```
very_common_vote_average(Movie_1,Movie_2):-
  vote_average(Movie_1,X),
  vote_average(Movie_2,Y),
  Movie_1 \= Movie_2,
  Dif is X-Y,
  Dif_abs is abs(Dif),
  Dif_abs =< 0.2.
```
```
partly_common_vote_average(Movie_1,Movie_2):-
  vote_average(Movie_1,X),
  vote_average(Movie_2,Y),
  Movie_1 \= Movie_2,
  Dif is X-Y,
  Dif_abs is abs(Dif),
  Dif_abs =< 0.5.
```

> Κοινό πλήθος ψήφων

```
very_common_num_voted(Movie_1,Movie_2):-
  num_voted_users(Movie_1,X),
  num_voted_users(Movie_2,Y),
  Movie_1 \= Movie_2,
  Dif is X-Y,
  Dif_abs is abs(Dif),
  Dif_abs =< 500.  
```
```
partly_common_num_voted(Movie_1,Movie_2):-
  num_voted_users(Movie_1,X),
  num_voted_users(Movie_2,Y),
  Movie_1 \= Movie_2,
  Dif is X-Y,
  Dif_abs is abs(Dif),
  Dif_abs =< 1500. 
```

> Κοινή χώρα

```
same_country(Movie_1,Movie_2):-
  country(Movie_1,X),
  country(Movie_2,X),
  Movie_1 \= Movie_2.
```

Τους παραπάνω κανόνες αποθηκεύουμε στο αρχείο "movie_metadata_rules.pl".

##Εισαγωγή δεδομένων και prolog

Αρχικά πραγματοποιούμε τα απαραίτητα imports και τις απαραίτητες εγκαταστάσεις βιβλιοθηκών:

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

import os
path="/content/drive/My Drive/Colab Notebooks/artificial_intelligence/second/" # με ευκόλο τρόπο μπορείτε να αλλάξετε το path που είναι αποθηκευμένα τα αρχεία σας

Mounted at /content/drive


In [ ]:
%%capture
#install swi-prolog
!sudo apt-get install software-properties-common
!sudo apt-add-repository ppa:swi-prolog
!sudo apt-get update
!sudo apt-get install swi-prolog
#install pyswip
!pip install pyswip

In [ ]:
from pyswip import Prolog
# Η βιβλιοθήκη pandas είναι χρήσιμη για την εργασία με τέτοια δεδομένα
import pandas as pd 

Στην συνέχεια διαβάζουμε το αρχείο movies_metadata.csv ως panda frame και θέτουμς την τιμή 'UNK' στα κελιά με nan τιμές:

In [ ]:
# Διάβασμα του αρχείου 'movie_metadata.csv' 
data = pd.read_csv("/content/drive/My Drive/Colab Notebooks/artificial_intelligence/second/movies_metadata.csv") 
#Στο csv υπαρχούν κελία με nan τιμές
#Στις θέσεις αυτές βάζουμε 'UNK' πράγμα που  κάνουμε με την παρακάτω συνάρτηση
data.fillna("UNK", inplace=True)
# Preview the first 5 lines of the loaded data 
data.head()

Unnamed: 0     budget  ...      actor_2_name       actor_3_name
0           0  237000000  ...  Sigourney Weaver       Stephen Lang
1           1  300000000  ...   Keira Knightley  Stellan Skarsgård
2           2  245000000  ...       Léa Seydoux      Ralph Fiennes
3           3  250000000  ...       Gary Oldman      Anne Hathaway
4           4  260000000  ...   Samantha Morton       Willem Dafoe

[5 rows x 27 columns]

##Ένταξη κατηγορημάτων στην Prolog

Στο σημείο αυτό θα εντάξουμε τα κατηγορήματα που δημιουργήσαμε νωρίτερα στην Prolog:

In [ ]:
def clean_text(text):                   # συνάρτηση που επεξεργάζειται strings πραγματοποιώντας ορισμένες απαραίτητες αντικαταστάσεις χαρακτήρων
  text = text.replace(u'\xa0', u'')
  text = text.replace(u"'", u'')
  return text

In [ ]:
import ast
#create World
#Ορίζουμε τον κόσμο μας
prolog = Prolog()

#Για κάθε row του πίνακα φτιάχνουμε τα κατηγορήματα που θέλουμε να αποθηκέυσουμε
#αρχικά σε μια λίστα με το όνομα literals
literals = []
movie_score = {}
for row in data.itertuples(index=True, name='Pandas'):
  
  #___simple attributes___:                                      # διαβάζουμε όλες τις στείλες του πίνακα εκτός αυτών οι οποίες δεν έχουν στα κελιά τους μία απλή τιμή π.χ. plot_keywords, και καλούμε για αυτά την clean_text

  movie_title = clean_text(getattr(row, 'movie_title'))
  budget = clean_text(str(getattr(row,'budget')))
  homepage = clean_text(getattr(row,'homepage'))
  id = clean_text(str(getattr(row,'id')))
  language = clean_text(getattr(row,'language'))
  original_title = clean_text(getattr(row,'original_title'))
  overview = clean_text(getattr(row,'overview'))
  popularity= clean_text(str(getattr(row,'popularity')))
  release_date= clean_text(getattr(row,'release_date'))
  gross= clean_text(str(getattr(row,'gross')))
  duration= clean_text(str(getattr(row,'duration')))
  status = clean_text(getattr(row,'status'))
  tagline = clean_text(getattr(row,'tagline'))
  movie_title = clean_text(getattr(row, 'movie_title'))
  vote_average= clean_text(str(getattr(row,'vote_average')))
  num_voted_users=clean_text(str(getattr(row,'num_voted_users')))
  title_year=clean_text(str(getattr(row,'title_year')))
  country = clean_text(getattr(row,'country'))   
  director= clean_text(getattr(row,'director_name'))
  actor_1_name = clean_text(getattr(row,'actor_1_name'))
  actor_2_name = clean_text(getattr(row,'actor_2_name'))
  actor_3_name = clean_text(getattr(row,'actor_3_name'))

  #insert into literals: distinction between strings and integers:                      # για την εισαγωγή των κανόνων στην prolog Θα πρέπει να υπάρχει διάκριση σε strings και integers καθώς στην πρώτη περίπτωση χρησιμοποιούμε τα σύμβολα: ' '

  #strings:

  if homepage !="UNK":
    literals.append("homepage('"+ movie_title +"','" + homepage + "')")
  if language !="UNK":
    literals.append("language('"+ movie_title +"','" + language + "')")
  if original_title !="UNK":
    literals.append("original_title('"+ movie_title +"','" + original_title + "')")
  if overview !="UNK":
    literals.append("overview('"+ movie_title +"','" + overview + "')")
  if release_date !="UNK":
    literals.append("release_date('"+ movie_title +"','" + release_date + "')")
  if status !="UNK":
    literals.append("status('"+ movie_title +"','" + status + "')")
  if tagline !="UNK":
    literals.append("tagline('"+ movie_title +"','" + tagline + "')")
  if movie_title !="UNK":
    literals.append("movie_title('"+ movie_title +"','" + movie_title + "')")
  if country !="UNK":
    literals.append("country('"+ movie_title +"','" + country + "')")
  if director !="UNK":
    literals.append("director_name('"+ movie_title +"','" + director + "')")
  if actor_1_name !="UNK":
    literals.append("actor_1_name('"+ movie_title +"','" + actor_1_name + "')")
  if actor_2_name !="UNK":
    literals.append("actor_2_name('"+ movie_title +"','" + actor_2_name + "')")
  if actor_3_name !="UNK":
    literals.append("actor_3_name('"+ movie_title +"','" + actor_3_name + "')")

  #integers:

  if budget !="UNK":
    literals.append("budget('"+ movie_title +"'," + budget + ")")
  if id !="UNK":
    literals.append("id('"+ movie_title +"'," + id + ")")
  if popularity !="UNK":
    literals.append("popularity('"+ movie_title +"'," + popularity + ")")
  if gross !="UNK":
    literals.append("gross('"+ movie_title +"'," + gross + ")")
  if duration !="UNK":
    literals.append("duration('"+ movie_title +"'," + duration + ")")
  if vote_average !="UNK":
    literals.append("vote_average('"+ movie_title +"'," + vote_average + ")")
  if num_voted_users !="UNK":
    literals.append("num_voted_users('"+ movie_title +"'," + num_voted_users + ")")
  if title_year !="UNK":
    literals.append("title_year('"+ movie_title +"'," + title_year + ")")

    
 #___list of attributes seperated by "|"___:                                                  # διαχειριζόμαστε τώρα τις στήλες με πολλαπλές τιμές

  for genre in getattr(row, 'genres').split("|"):
    literals.append("genre('"+ movie_title +"','"+ clean_text(genre) +"')")
  for keyword in getattr(row, 'plot_keywords').split("|"):
    literals.append("plot_keyword('"+ movie_title +"','"+ clean_text(keyword) +"')")

 #___list of attributes in dictionary form___:                                                # διαχειριζόμαστε τώρα τις στήλες με τιμές σε μορφή λεξικού

  country_dict=ast.literal_eval(getattr(row, 'production_countries'))   #convert string to dictionary list
  if (country_dict!=[]):
    for c in country_dict:
      literals.append("production_country('"+movie_title+"','"+clean_text(c['iso_3166_1'])+"','"+clean_text(c['name'])+"')")

  company_dict=ast.literal_eval(getattr(row, 'production_companies')) 
  if (company_dict!=[]):
    for c in company_dict:
      literals.append("production_company('"+movie_title+"','"+clean_text(c['name'])+"',"+clean_text(str(c['id']))+")")

  spoken_languages_dict=ast.literal_eval(getattr(row, 'spoken_languages')) 
  if (spoken_languages_dict!=[]):
    for s in spoken_languages_dict:
      literals.append("spoken_language('"+movie_title+"','"+clean_text(s['iso_639_1'])+"','"+clean_text(s['name'])+"')")

#Η Prolog θέλει τα κατηγορήματά της με την σειρά 
literals.sort()
for literal in literals:
  prolog.assertz(literal)               # εισάγουμε τα κατηγορήματα στην Prolog

##Ένταξη κανόνων στην Prolog

In [ ]:
prolog.consult(path + "movie_metadata_rules.pl")   # ορίζουμε η  Prolog να λάβει υπ' όψιν τους κανόνες που περιλαμβάνει το συγκεκριμένο αρχείο

##Επιλεκτικές δοκιμές για κάποιους από τους κανόνες

In [ ]:
q = prolog.query("four_common_genre('Avatar', Y)")
s = set()
for soln in q:
  m = soln['Y'] 
  if m not in s:
    s.add(soln['Y'])
print(s)

{'The Fifth Element', 'X-Men: Days of Future Past', 'Thunderbirds', 'Final Fantasy: The Spirits Within', 'Jimmy Neutron: Boy Genius', 'Mystery Men', 'Blade: Trinity', 'Race to Witch Mountain', 'Spawn', 'Superman II', 'Underworld: Rise of the Lycans', 'Star Wars: Clone Wars: Volume 1', 'Superman III', 'The Covenant', 'Superman', 'Man of Steel', 'Jupiter Ascending', 'Small Soldiers', 'Teenage Mutant Ninja Turtles', 'Beastmaster 2: Through the Portal of Time', 'The Shadow', 'Superman Returns', 'Dragonball Evolution', 'Suicide Squad', 'Teenage Mutant Ninja Turtles III', 'Batman', 'Sheena', 'The Wolverine'}


In [ ]:
q = prolog.query("five_same_plot('Avengers: Age of Ultron', Y)")
s = set()
for soln in q:
  m = soln['Y'] 
  if m not in s:
    s.add(soln['Y'])
print(s)

{'Iron Man 3', 'Thor: The Dark World', 'Fantastic 4: Rise of the Silver Surfer', 'Captain America: The Winter Soldier', 'Captain America: Civil War', 'Ant-Man', 'The Avengers', 'Thor', 'Captain America: The First Avenger'}


In [ ]:
q = prolog.query("common_director('Spectre', Y)")
s = set()
for soln in q:
  m = soln['Y'] 
  if m not in s:
    s.add(soln['Y'])
print(s)

{'Away We Go', 'Revolutionary Road', 'Jarhead', 'American Beauty', 'Skyfall', 'Road to Perdition'}


In [ ]:
q = prolog.query("some_same_actors('Toy Story 3', Y)")
s = set()
for soln in q:
  m = soln['Y'] 
  if m not in s:
    s.add(soln['Y'])
print(s)

{'Toy Story 2'}


In [ ]:
q = prolog.query("one_same_actor('Titanic', Y)")
s = set()
for soln in q:
  m = soln['Y'] 
  if m not in s:
    s.add(soln['Y'])
print(s)

{'Django Unchained', 'The Quick and the Dead', 'The Believer', 'Whats Eating Gilbert Grape', 'Blue Car', 'In the Valley of Elah', 'Marvins Room', 'Celebrity', 'House of Sand and Fog'}


In [ ]:
q = prolog.query("same_language('Frances Ha', Y)")
s = set()
for soln in q:
  m = soln['Y'] 
  if m not in s:
    s.add(soln['Y'])
print(s)

{'All Good Things', 'The Last Exorcism', 'Winter Passing', 'Fury', 'DysFunktional Family', 'Jimmy and Judy', 'You Dont Mess with the Zohan', 'Serendipity', 'Sphere', 'House of Wax', 'The Rainmaker', 'Beverly Hills Cop III', 'Limbo', 'The Broken Hearts Club: A Romantic Comedy', 'Spotlight', 'Jawbreaker', 'Pay It Forward', 'Chicken Tikka Masala', 'Gridiron Gang', 'The Omega Code', 'Beer League', 'Aladdin', 'Any Given Sunday', 'National Lampoons Loaded Weapon 1', 'Hocus Pocus', 'The Silence of the Lambs', 'Renaissance Man', 'The Muse', 'H.', 'Notting Hill', 'That Thing You Do!', 'The Book of Life', 'Kicking & Screaming', 'The Ghost Writer', 'The Railway Man', 'Ultraviolet', 'From Russia with Love', 'Meet Dave', 'Monsters, Inc.', 'Muppets from Space', 'A Thin Line Between Love and Hate', 'Steamboy', 'Brave', 'Penguins of Madagascar', 'Maniac', 'Ondine', 'The American President', 'Unstoppable', 'Amigo', 'The Switch', 'Enough', 'Contraband', 'Yoga Hosers', 'The Perfect Storm', 'The Other End

In [ ]:
# for a black and white movie
q = prolog.query("black_and_white('Frances Ha', Y)")
s = set()
for soln in q:
  m = soln['Y'] 
  if m not in s:
    s.add(soln['Y'])
print(s)

{'Creative Control', 'Frankenweenie', 'The Giver', 'Renaissance', 'The Phantom of the Opera'}


In [ ]:
# for a movie with color
q = prolog.query("black_and_white('Avatar', Y)")
s = set()
for soln in q:
  m = soln['Y'] 
  if m not in s:
    s.add(soln['Y'])
print(s)

{'All Good Things', 'The Last Exorcism', 'Winter Passing', 'Fury', 'DysFunktional Family', 'Jimmy and Judy', 'The Exorcist', 'You Dont Mess with the Zohan', 'Serendipity', 'Sphere', 'House of Wax', 'The Rainmaker', 'Beverly Hills Cop III', 'Underdogs', 'Limbo', 'The Sticky Fingers of Time', 'The Broken Hearts Club: A Romantic Comedy', 'Spotlight', 'Jawbreaker', 'Pay It Forward', 'Rounders', 'The Cave', 'Chicken Tikka Masala', 'Gridiron Gang', 'See Spot Run', 'The Christmas Candle', 'The Omega Code', 'Oceans Eleven', 'Beer League', 'Aladdin', 'Any Given Sunday', 'National Lampoons Loaded Weapon 1', 'Hocus Pocus', 'The Silence of the Lambs', 'Amélie', 'Renaissance Man', 'The Muse', 'H.', 'The Reaping', 'Notting Hill', 'That Thing You Do!', 'The Book of Life', 'Kicking & Screaming', 'The Ghost Writer', 'The Railway Man', 'Ultraviolet', 'From Russia with Love', 'Meet Dave', 'Monsters, Inc.', 'Muppets from Space', 'A Thin Line Between Love and Hate', 'Steamboy', 'Brave', 'Penguins of Madaga

In [ ]:
q = prolog.query("two_common_production_studio('Man of Steel', Y)")
s = set()
for soln in q:
  m = soln['Y'] 
  if m not in s:
    s.add(soln['Y'])
print(s)

{'Lady in the Water', 'Pacific Rim', 'The Hangover Part II', 'The Prestige', 'Where the Wild Things Are', '300', 'Steel', 'Due Date', 'The Dark Knight', 'Inception', 'Wrath of the Titans', 'Batman v Superman: Dawn of Justice', '300: Rise of an Empire', 'Sucker Punch', 'Green Lantern', 'Legend of the Guardians: The Owls of GaHoole', 'The Town', 'Ninja Assassin', 'Batman Begins', 'The Hangover', 'Watchmen', 'Superman Returns', 'Jack the Giant Slayer', 'Suicide Squad', 'Interstellar', 'The Losers', 'We Are Marshall', 'The Dark Knight Rises', 'Clash of the Titans'}


In [ ]:
q = prolog.query("one_common_production_country('Skyfall', Y)")
s = set()
for soln in q:
  m = soln['Y'] 
  if m not in s:
    s.add(soln['Y'])
print(s)

{'All Good Things', 'The Last Exorcism', 'Winter Passing', 'Fury', 'DysFunktional Family', 'Jimmy and Judy', 'You Dont Mess with the Zohan', 'Serendipity', 'Sphere', 'House of Wax', 'The Rainmaker', 'Beverly Hills Cop III', 'The Broken Hearts Club: A Romantic Comedy', 'Spotlight', 'Jawbreaker', 'Pay It Forward', 'Rounders', 'The Cave', 'Chicken Tikka Masala', 'Gridiron Gang', 'See Spot Run', 'Oceans Eleven', 'Beer League', 'Aladdin', 'Any Given Sunday', 'National Lampoons Loaded Weapon 1', 'Hocus Pocus', 'The Silence of the Lambs', 'Renaissance Man', 'The Muse', 'H.', 'The Reaping', 'Notting Hill', 'That Thing You Do!', 'The Book of Life', 'The Ghost Writer', 'The Railway Man', 'Kicking & Screaming', 'Ultraviolet', 'From Russia with Love', 'Meet Dave', 'Monsters, Inc.', 'Muppets from Space', 'A Thin Line Between Love and Hate', 'Brave', 'Penguins of Madagascar', 'Maniac', 'Ondine', 'The American President', 'Unstoppable', 'Amigo', 'The Switch', 'Enough', 'Contraband', 'Yoga Hosers', 'T

In [ ]:
q = prolog.query("common_decade('Avatar', Y)")
s = set()
for soln in q:
  m = soln['Y'] 
  if m not in s:
    s.add(soln['Y'])
print(s)

{'Winter Passing', 'DysFunktional Family', 'Jimmy and Judy', 'You Dont Mess with the Zohan', 'Serendipity', 'House of Wax', 'The Broken Hearts Club: A Romantic Comedy', 'Pay It Forward', 'The Cave', 'Chicken Tikka Masala', 'Gridiron Gang', 'See Spot Run', 'Oceans Eleven', 'Beer League', 'Amélie', 'Kicking & Screaming', 'Ultraviolet', 'Meet Dave', 'Monsters, Inc.', 'Steamboy', 'Ondine', 'Enough', 'The Perfect Storm', 'The Other End of the Line', 'Three Kingdoms: Resurrection of the Dragon', 'Birthday Girl', 'The Game of Their Lives', 'Marie Antoinette', 'Lara Croft Tomb Raider: The Cradle of Life', 'Training Day', 'Stuart Little 2', 'The Wash', 'Mutual Appreciation', 'The Business of Strangers', 'Shine a Light', 'Kingdom of Heaven', 'Dogtooth', 'Underworld: Rise of the Lycans', 'Married Life', 'Once', 'Vicky Cristina Barcelona', 'Orphan', 'Monster House', 'Secondhand Lions', 'Dead Snow', 'Driven', 'Bamboozled', 'Fish Tank', 'Slither', 'Love & Basketball', 'The Last King of Scotland', 'T

In [ ]:
q = prolog.query("partly_common_budget('Harry Potter and the Half-Blood Prince', Y)")
s = set()
for soln in q:
  m = soln['Y'] 
  if m not in s:
    s.add(soln['Y'])
print(s)

{'The Chronicles of Narnia: Prince Caspian', 'Tangled', 'X-Men: Days of Future Past', 'Terminator 3: Rise of the Machines', 'Iron Man 2', 'Spider-Man 3', 'Spectre', 'World War Z', 'King Kong', 'Alice in Wonderland', 'Skyfall', 'Avengers: Age of Ultron', 'Transformers: Age of Extinction', 'Pirates of the Caribbean: Dead Mans Chest', 'A Christmas Carol', 'Iron Man 3', 'Battleship', 'The Lone Ranger', 'Titanic', 'Men in Black 3', 'The Hobbit: An Unexpected Journey', 'John Carter', 'Man of Steel', 'Batman v Superman: Dawn of Justice', 'Pirates of the Caribbean: At Worlds End', 'Monsters University', 'Green Lantern', 'The Amazing Spider-Man', 'The Amazing Spider-Man 2', '2012', 'Avatar', 'Spider-Man 2', 'Cars 2', 'Quantum of Solace', 'The Avengers', 'Superman Returns', 'Oz: The Great and Powerful', 'The Hobbit: The Battle of the Five Armies', 'The Hobbit: The Desolation of Smaug', 'The Dark Knight Rises', 'Toy Story 3', 'Captain America: Civil War', 'Robin Hood', 'X-Men: The Last Stand', 'T

In [ ]:
q = prolog.query("common_duration('Avatar', Y)")
s = set()
for soln in q:
  m = soln['Y'] 
  if m not in s:
    s.add(soln['Y'])
print(s)

{'Fury', 'The Equalizer', 'Spectre', 'Sphere', 'For Love of the Game', 'The Rainmaker', 'Hoffa', 'Skyfall', 'Across the Universe', 'Barry Lyndon', 'Any Given Sunday', 'Boogie Nights', 'Das Boot', 'Princess Mononoke', 'The Blues Brothers', 'Sling Blade', 'The English Patient', 'Fugly', 'The White Ribbon', 'Dreamgirls', 'The Patriot', 'Ta Ra Rum Pum', 'Legends of the Fall', 'For Colored Girls', 'Hidalgo', 'Furious 7', 'The Last Temptation of Christ', 'Hook', 'Avengers: Age of Ultron', 'Little White Lies', 'Restless', 'Gone Girl', 'Pulp Fiction', 'Mr. Turner', 'Speed Racer', 'Cold Mountain', 'The Lone Ranger', 'Robin Hood: Prince of Thieves', 'Kingdom of Heaven', 'The Hobbit: An Unexpected Journey', 'Finding Forrester', 'Barneys Version', 'Margaret', 'Rosemarys Baby', 'Oliver!', 'Noah', 'The Lord of the Rings: The Two Towers', 'Mr. Hollands Opus', 'Open Range', 'Bamboozled', 'The Place Beyond the Pines', 'The Nuns Story', 'Interstellar', 'The Postman', 'Riding in Cars with Boys', 'Amadeus

In [ ]:
q = prolog.query("two_common_spoken_language('Spectre', Y)")
s = set()
for soln in q:
  m = soln['Y'] 
  if m not in s:
    s.add(soln['Y'])
print(s)

{'The Christmas Bunny', 'Butch Cassidy and the Sundance Kid', 'Pontypool', 'What Women Want', 'Dave Chappelles Block Party', 'Under the Tuscan Sun', 'Raising Victor Vargas', 'The Rocker', 'Simply Irresistible', 'To Rome with Love', 'Underdogs', 'Hoffa', 'The Broken Hearts Club: A Romantic Comedy', 'Barry Lyndon', 'The Omega Code', 'Oceans Eleven', 'Mad Hot Ballroom', 'By the Sea', 'Chasing Liberty', 'Safe House', 'Renaissance Man', 'Young Frankenstein', 'Taken', 'Das Boot', 'Crazy/Beautiful', 'Notting Hill', 'Slumdog Millionaire', 'Kicking & Screaming', 'The English Patient', 'The White Ribbon', 'Ondine', 'The Sweet Hereafter', 'Dirty Pretty Things', 'Boogeyman', 'Dinner Rush', 'Four Single Fathers', 'Clean', 'Instructions Not Included', 'Yoga Hosers', 'Blazing Saddles', 'The Lost Medallion: The Adventures of Billy Stone', 'The Big Wedding', 'The Vatican Exorcisms', 'No Country for Old Men', 'Inspector Gadget', 'Brüno', 'Marie Antoinette', 'Training Day', 'U.F.O.', 'What Just Happened'

In [ ]:
q = prolog.query("partly_common_vote_average('Avatar', Y)")
s = set()
for soln in q:
  m = soln['Y'] 
  if m not in s:
    s.add(soln['Y'])
print(s)

{'Winter Passing', 'Fury', 'The Rainmaker', 'Limbo', 'Pay It Forward', 'Rounders', 'Oceans Eleven', 'Aladdin', 'Any Given Sunday', 'Notting Hill', 'That Thing You Do!', 'The Book of Life', 'The Ghost Writer', 'The Railway Man', 'From Russia with Love', 'Monsters, Inc.', 'Brave', 'Instructions Not Included', 'Avengers: Age of Ultron', 'Training Day', 'Charlie St. Cloud', 'Shine a Light', 'Ever After: A Cinderella Story', 'Dogtooth', 'The Hobbit: An Unexpected Journey', 'Once', 'Vicky Cristina Barcelona', 'It Happened One Night', 'Orphan', 'The Fighter', 'Secondhand Lions', 'Fish Tank', 'Love & Basketball', 'The Nuns Story', 'Sleeper', 'The Last King of Scotland', 'Open Secret', 'Back to the Future Part III', 'The Last Castle', 'The Road to El Dorado', 'Inside Man', 'Vaalu', 'The Fifth Element', 'Sympathy for Lady Vengeance', 'Clerks II', 'Kung Fu Hustle', 'Mad Max 2: The Road Warrior', 'Kids', 'A Madea Christmas', 'Henry V', 'Apocalypto', 'Max', 'Saint Ralph', 'Face/Off', 'Things to Do 

# **Μέρος 2: Recommendation System με βάση μόνο τα χαρακτηριστικά των ταινιών.**



##Εξήγηση κατηγορημάτων find_similar

Στο σημείο αυτό, με βάση τους κανόνες που κατασκευάστηκαν στο Μέρος 1,  κατασκευάζουμε κατηγορήματα για την ομοιότητα ταινιών. Συγκεκριμένα κατασκευάζουμε 10 κατηγορήματα, find_similar_1 έως και find_similar_10, με το find_similar_10 να βρίσκει ταινίες πιο όμοιες με την ζητούμενη από το find_similar_9 κοκ, και καταλήγωντας στο find_similar_1 που επιστρέφει τις λιγότερες όμοιες ταινίες με την ζητούμενη. Τα κατηγορήματα περιέχονται στο αρχείο find_similar.pl το οποίο ορίζουμε να συμβουλεύεται η Prolog παρακάτω.

Σημειώνεται ότι κάθε κατηγόρημα find similar αποτελείται από κάποια, συνήθως 3, if-else clauses (συνδεδεμένα μεταξύ τους με λογικό OR), σε καθένα από τα οποία ορίζεται ένα σύνολο κανόνων που πρέπει να ικανοποιούνται ταυτόχρονα. Όσο ανεβαίνουμε επίπεδα ομοιότητας σε καθένα από αυτά τα if-else clauses προστίθενται διαφορετικοί κανόνες με αποτέλεσμα το σύνολο κανόνων του κάθε clause να γίνεται ολοένα και πιο αυστηρό. Στο τέλος, στο επίπεδο 10 που είναι και οι πιο όμοιες ταινίες όλοι οι κανόνες που έχουν οριστεί στα επιμέρους if-else clauses ενωποιούνται, δηλαδή ενώνονται μεταξύ τους με λογικά AND, και πρέπει να ικανοποιούνται όλοι ταυτόχρονα.

Αξίζει να αναφερθεί ότι αρχικά είχαν οριστεί κατηγορήματα find similar που εμπεριείχαν κανόνες συνδεδεμένους μόνο με λογικά AND. Άρα ξεκινώντας από το επίπεδο ομοιότητας 1 με έναν κανόνα και ανεβαίνοντας επίπεδα ομοιότητας απλά προσθέτονταν κανόνες οι οποίοι θα έπρεπε να ικανοποιούνται όλοι ταυτόχρονα. Ωστόσο στην πορεία αντιληφθήκαμε ότι η συγκεκριμένη λογική δεν ήταν βέλτιστη, αφού για να χαρακτηριστεί μια ταινάι ως όμοια με μια αλλή δεν είναι απαραίτητο να ικανοποιεί μόνο ένα σύνολο κανόνων. Για παράδειγμα, δύο ταινίες θα μπορούσαν να μοιάζουν γιατί έχουν τον ίδιο πρωταγωνιστή και δύο άλλες γιατί έχουν τον ίδιο σκηνοθέτη, χωρίς να υπάρχει αντικειμενική διάκριση για το σε ποια περίπτωση οι ταινίες μοιάζουν πιο πολύ.

Έχοντας αυτό στο μυαλό μας τροποποιήσαμε τους αρχικούς μας πολύ αυστηρούς, μονοδιάστατους κανόνες find similar και δημιουργήσαμε, όσο καλύτερα μπορούσαμε, την δομή με τα if-else clauses που περιγράφηκε παραπάνω. Κάνοντας την αλλαγή αυτή παρατηρήσαμε καλύτερη κλιμάκωση στα επίπεδα find similar, δηλαδή για μια ταινία βρίσκονται όμοιες της μέχρι και μεγάλα επίπεδα ομοιότητας, κάτι το οποίο δεν συνέβαινε με την προηγούμενη, αυστηρή δομή που είχε επιλεγεί στην αρχή.

Γίνεται μια ***γενική περιγραφή των κριτηρίων ομοιότητας*** των ταινιών που χρησιμοποιήθηκαν σε κάθε κατηγόρημα find_similar_Χ: 

> **find_similar_1**

*   1 κοινό θέμα (μια λέξη genre κοινή)

    **ή**

*   Ελάχιστα ίδια υπόθεση (1 λέξη κλειδί plot_keyword κοινή)
    
    **ή**

*   Έχουν κυκλοφορήσει την ίδια δεκαετία

> **find_similar_2**

*   1 κοινό θέμα (μια λέξη genre κοινή)
*   Ελάχιστα ίδια υπόθεση (1 λέξη κλειδί plot_keyword κοινή)

    **ή**

*   Ελάχιστα ίδια υπόθεση (1 λέξη κλειδί plot_keyword κοινή)
*   Έχουν κυκλοφορήσει την ίδια δεκαετία
    
    **ή**

*   Έχουν κυκλοφορήσει την ίδια δεκαετία
*   1 κοινό θέμα (μια λέξη genre κοινή)


> **find_similar_3**

*   1 κοινό θέμα (μια λέξη genre κοινή)
*   Ελάχιστα ίδια υπόθεση (2 λέξεις κλειδιά plot_keyword κοινές)
*   Μια ομιλούμενη γλώσσα κοινή (one_common_spoken_language)

    **ή**

*   Ελάχιστα ίδια υπόθεση (2 λέξεις κλειδιά plot_keyword κοινές)
*   Έχουν κυκλοφορήσει την ίδια δεκαετία
*   Μια ομιλούμενη γλώσσα κοινή (one_common_spoken_language)

    **ή**

*   Έχουν κυκλοφορήσει την ίδια δεκαετία
*   1 κοινό θέμα (μια λέξη genre κοινή)
*   Μια ομιλούμενη γλώσσα κοινή (one_common_spoken_language)



> **find_similar_4**

*   1 κοινό θέμα (μια λέξη genre κοινή)
*   Ελάχιστα ίδια υπόθεση (2 λέξεις κλειδιά plot_keyword κοινές)
*   Μια ομιλούμενη γλώσσα κοινή (one_common_spoken_language)
*   Ίδιο χρώμα εικόνας (είτε όλες ασπρόμαυρες είτε όλες έγχρωμες)

    **ή**

*   Ελάχιστα ίδια υπόθεση (2 λέξεις κλειδιά plot_keyword κοινές)
*   Έχουν κυκλοφορήσει την ίδια δεκαετία
*   Μια ομιλούμενη γλώσσα κοινή (one_common_spoken_language)
*   Ίδιο χρώμα εικόνας (είτε όλες ασπρόμαυρες είτε όλες έγχρωμες)

    **ή**

*   Έχουν κυκλοφορήσει την ίδια δεκαετία
*   1 κοινό θέμα (μια λέξη genre κοινή)
*   Μια ομιλούμενη γλώσσα κοινή (one_common_spoken_language)
*   Ίδιο χρώμα εικόνας (είτε όλες ασπρόμαυρες είτε όλες έγχρωμες)


> **find_similar_5**

*   1 κοινό θέμα (μια λέξη genre κοινή)
*   Αρκετά ίδια υπόθεση (3 λέξεις κλειδιά plot_keyword κοινές)
*   Μια ομιλούμενη γλώσσα κοινή (one_common_spoken_language)
*   Ίδιο χρώμα εικόνας (είτε όλες ασπρόμαυρες είτε όλες έγχρωμες)

    **ή**

*   Ελάχιστα ίδια υπόθεση (2 λέξεις κλειδιά plot_keyword κοινές)
*   Έχουν κυκλοφορήσει την ίδια δεκαετία
*   Μια ομιλούμενη γλώσσα κοινή (one_common_spoken_language)
*   Ίδιο χρώμα εικόνας (είτε όλες ασπρόμαυρες είτε όλες έγχρωμες)
*   Ένας ίδιος ηθοποιός
*   1 κοινό θέμα (μια λέξη genre κοινή)

    **ή**

*   Έχουν κυκλοφορήσει την ίδια δεκαετία
*   1 κοινό θέμα (μια λέξη genre κοινή)
*   Μια ομιλούμενη γλώσσα κοινή (one_common_spoken_language)
*   Ίδιο χρώμα εικόνας (είτε όλες ασπρόμαυρες είτε όλες έγχρωμες)
*   Αρκετά ίδια υπόθεση (3 λέξεις κλειδιά plot_keyword κοινές)



> **find_similar_6**

*   1 κοινό θέμα (μια λέξη genre κοινή)
*   Αρκετά ίδια υπόθεση (3 λέξεις κλειδιά plot_keyword κοινές)
*   Μια ομιλούμενη γλώσσα κοινή (one_common_spoken_language)
*   Ίδιο χρώμα εικόνας (είτε όλες ασπρόμαυρες είτε όλες έγχρωμες)
*   Ίδιο σκηνοθέτη
*   Έχουν κυκλοφορήσει την ίδια δεκαετία 

    **ή**

*   Ελάχιστα ίδια υπόθεση (2 λέξεις κλειδιά plot_keyword κοινές)
*   Έχουν κυκλοφορήσει την ίδια δεκαετία
*   Μια ομιλούμενη γλώσσα κοινή (one_common_spoken_language)
*   Ίδιο χρώμα εικόνας (είτε όλες ασπρόμαυρες είτε όλες έγχρωμες)
*   Ένας ίδιος ηθοποιός
*   1 κοινό θέμα (μια λέξη genre κοινή)
*   Σχετικά ίδια βαθμολογία

    **ή**

*   Έχουν κυκλοφορήσει την ίδια δεκαετία
*   1 κοινό θέμα (μια λέξη genre κοινή)
*   Μια ομιλούμενη γλώσσα κοινή (one_common_spoken_language)
*   Ίδιο χρώμα εικόνας (είτε όλες ασπρόμαυρες είτε όλες έγχρωμες)
*   Αρκετά ίδια υπόθεση (3 λέξεις κλειδιά plot_keyword κοινές)
*   Ένας ίδιος ηθοποιός



> **find_similar_7**

*   1 κοινό θέμα (μια λέξη genre κοινή)
*   Αρκετά ίδια υπόθεση (3 λέξεις κλειδιά plot_keyword κοινές)
*   Μια ομιλούμενη γλώσσα κοινή (one_common_spoken_language)
*   Ίδιο χρώμα εικόνας (είτε όλες ασπρόμαυρες είτε όλες έγχρωμες)
*   Ίδιο σκηνοθέτη
*   Έχουν κυκλοφορήσει την ίδια δεκαετία
*   Ένας ίδιος ηθοποιός 

    **ή**

*   Ελάχιστα ίδια υπόθεση (2 λέξεις κλειδιά plot_keyword κοινές)
*   Έχουν κυκλοφορήσει την ίδια δεκαετία
*   Μια ομιλούμενη γλώσσα κοινή (one_common_spoken_language)
*   Ίδιο χρώμα εικόνας (είτε όλες ασπρόμαυρες είτε όλες έγχρωμες)
*   Ένας ίδιος ηθοποιός
*   1 κοινό θέμα (μια λέξη genre κοινή)
*   Σχετικά ίδια βαθμολογία
*   Ίδιο σκηνοθέτη

    **ή**

*   Έχουν κυκλοφορήσει την ίδια δεκαετία
*   1 κοινό θέμα (μια λέξη genre κοινή)
*   Μια ομιλούμενη γλώσσα κοινή (one_common_spoken_language)
*   Ίδιο χρώμα εικόνας (είτε όλες ασπρόμαυρες είτε όλες έγχρωμες)
*   Αρκετά ίδια υπόθεση (3 λέξεις κλειδιά plot_keyword κοινές)
*   Ένας ίδιος ηθοποιός
*   Σχετικά κοινό budget


> **find_similar_8**

*   1 κοινό θέμα (μια λέξη genre κοινή)
*   Αρκετά ίδια υπόθεση (3 λέξεις κλειδιά plot_keyword κοινές)
*   Μια ομιλούμενη γλώσσα κοινή (one_common_spoken_language)
*   Ίδιο χρώμα εικόνας (είτε όλες ασπρόμαυρες είτε όλες έγχρωμες)
*   Ίδιο σκηνοθέτη
*   Έχουν κυκλοφορήσει την ίδια δεκαετία
*   Ένας ίδιος ηθοποιός 
*   Έναν ίδιο παραγωγό
*   Σχετικά ίδια βαθμολογία

    **ή**

*   Ελάχιστα ίδια υπόθεση (2 λέξεις κλειδιά plot_keyword κοινές)
*   Έχουν κυκλοφορήσει την ίδια δεκαετία
*   Μια ομιλούμενη γλώσσα κοινή (one_common_spoken_language)
*   Ίδιο χρώμα εικόνας (είτε όλες ασπρόμαυρες είτε όλες έγχρωμες)
*   Ένας ίδιος ηθοποιός
*   1 κοινό θέμα (μια λέξη genre κοινή)
*   Σχετικά ίδια βαθμολογία
*   Ίδιο σκηνοθέτη
*   Έναν ίδιο παραγωγό
*   Σχετικά κοινό budget

    **ή**

*   Έχουν κυκλοφορήσει την ίδια δεκαετία
*   1 κοινό θέμα (μια λέξη genre κοινή)
*   Μια ομιλούμενη γλώσσα κοινή (one_common_spoken_language)
*   Ίδιο χρώμα εικόνας (είτε όλες ασπρόμαυρες είτε όλες έγχρωμες)
*   Αρκετά ίδια υπόθεση (3 λέξεις κλειδιά plot_keyword κοινές)
*   Ένας ίδιος ηθοποιός
*   Σχετικά κοινό budget
*   Ίδιο σκηνοθέτη
*   Έναν ίδιο παραγωγό


> **find_similar_9**

*   2 κοινά θέματα (2 λέξεις genre κοινές)
*   Αρκετά ίδια υπόθεση (3 λέξεις κλειδιά plot_keyword κοινές)
*   Μια ομιλούμενη γλώσσα κοινή (one_common_spoken_language)
*   Ίδιο χρώμα εικόνας (είτε όλες ασπρόμαυρες είτε όλες έγχρωμες)
*   Ίδιο σκηνοθέτη
*   Έχουν κυκλοφορήσει την ίδια δεκαετία
*   Ένας ίδιος ηθοποιός 
*   Έναν ίδιο παραγωγό
*   Σχετικά ίδια βαθμολογία
*   Παρόμοια διάρκεια
*   Ίδια γλώσσα
*   Σχετικά κοινό budget
*   Σχετικά κοινή δημοφιλία

    **ή**

*   Ελάχιστα ίδια υπόθεση (2 λέξεις κλειδιά plot_keyword κοινές)
*   Έχουν κυκλοφορήσει την ίδια δεκαετία
*   Μια ομιλούμενη γλώσσα κοινή (one_common_spoken_language)
*   Ίδιο χρώμα εικόνας (είτε όλες ασπρόμαυρες είτε όλες έγχρωμες)
*   Ένας ίδιος ηθοποιός
*   2 κοινά θέματα (2 λέξεις genre κοινές)
*   Σχετικά ίδια βαθμολογία
*   Ίδιο σκηνοθέτη
*   Έναν ίδιο παραγωγό
*   Σχετικά κοινό budget
*   Παρόμοια διάρκεια
*   Ίδια γλώσσα
*   Σχετικά κοινή δημοφιλία
*   Σχετικά ίδια κέρδη

    **ή**

*   Έχουν κυκλοφορήσει την ίδια δεκαετία
*   2 κοινά θέματα (2 λέξεις genre κοινές)
*   Μια ομιλούμενη γλώσσα κοινή (one_common_spoken_language)
*   Ίδιο χρώμα εικόνας (είτε όλες ασπρόμαυρες είτε όλες έγχρωμες)
*   Αρκετά ίδια υπόθεση (3 λέξεις κλειδιά plot_keyword κοινές)
*   Ένας ίδιος ηθοποιός
*   Σχετικά κοινό budget
*   Ίδιο σκηνοθέτη
*   Έναν ίδιο παραγωγό
*   Ίδια γλώσσα
*   Σχετικά κοινή δημοφιλία
*   Σχετικά ίδια βαθμολογία
*   Σχετικά ίδια κέρδη


> **find_similar_10**

*   Ίδια υπόθεση (4 λέξεις κλειδιά plot_keyword κοινές)
*   2 κοινά θέματα (2 λέξεις genre κοινές)
*   Μια ομιλούμενη γλώσσα κοινή (one_common_spoken_language)
*   Ίδιο χρώμα εικόνας (είτε όλες ασπρόμαυρες είτε όλες έγχρωμες)
*   Ίδιο σκηνοθέτη
*   Έχουν κυκλοφορήσει την ίδια δεκαετία
*   Μερικούς ίδιους ηθοποιούς (2 από τους 3)
*   1 ίδιο παραγωγό
*   Πολύ ίδια βαθμολογία
*   Παρόμοια διάρκεια
*   Ίδια γλώσσα
*   Πολύ κοινό budget
*   Πολύ κοινή δημοφιλία
*   Σχετικά ίδια κέρδη
*   2 ίδιες χώρες παραγωγής
*   Σχετικά ίδιο αριθμό ψήφων



Πιο συγκεκριμένα τα κατηγορήματα  find_similar είναι τα παρακάτω: 

```
find_similar_1(Movie_1,Movie_2):-
      one_common_genre(Movie_1,Movie_2);
      common_decade(Movie_1,Movie_2);
      one_same_plot(Movie_1,Movie_2).
```

```
find_similar_2(Movie_1,Movie_2):-
      one_same_plot(Movie_1,Movie_2),
      one_common_genre(Movie_1,Movie_2);
      one_same_plot(Movie_1,Movie_2),
      common_decade(Movie_1,Movie_2);
      common_decade(Movie_1,Movie_2),
      one_common_genre(Movie_1,Movie_2).
```

```
find_similar_3(Movie_1,Movie_2):-
      two_same_plot(Movie_1,Movie_2),
      one_common_genre(Movie_1,Movie_2),
      one_common_spoken_language(Movie_1,Movie_2);
      two_same_plot(Movie_1,Movie_2),
      common_decade(Movie_1,Movie_2),
      one_common_spoken_language(Movie_1,Movie_2);
      common_decade(Movie_1,Movie_2),
      one_common_genre(Movie_1,Movie_2),
      one_common_spoken_language(Movie_1,Movie_2).
  ```
  ```
find_similar_4(Movie_1,Movie_2):-
      two_same_plot(Movie_1,Movie_2),
      one_common_genre(Movie_1,Movie_2),
      one_common_spoken_language(Movie_1,Movie_2),
      black_and_white(Movie_1,Movie_2);
      two_same_plot(Movie_1,Movie_2),
      common_decade(Movie_1,Movie_2),
      one_common_spoken_language(Movie_1,Movie_2),
      black_and_white(Movie_1,Movie_2);
      one_common_genre(Movie_1,Movie_2),
      common_decade(Movie_1,Movie_2),
      one_common_spoken_language(Movie_1,Movie_2),
      black_and_white(Movie_1,Movie_2).
  ```
  ```
find_similar_5(Movie_1,Movie_2):-
      three_same_plot(Movie_1,Movie_2),
      one_common_genre(Movie_1,Movie_2),
      one_common_spoken_language(Movie_1,Movie_2),
      black_and_white(Movie_1,Movie_2);
      two_same_plot(Movie_1,Movie_2),
      common_decade(Movie_1,Movie_2),
      one_common_spoken_language(Movie_1,Movie_2),
      black_and_white(Movie_1,Movie_2),
      one_same_actor(Movie_1,Movie_2),
      one_common_genre(Movie_1,Movie_2);
      three_same_plot(Movie_1,Movie_2),
      one_common_genre(Movie_1,Movie_2),
      common_decade(Movie_1,Movie_2),
      one_common_spoken_language(Movie_1,Movie_2),
      black_and_white(Movie_1,Movie_2).
  ```
  ```
find_similar_6(Movie_1,Movie_2):-
      three_same_plot(Movie_1,Movie_2),
      one_common_genre(Movie_1,Movie_2),
      one_common_spoken_language(Movie_1,Movie_2),
      black_and_white(Movie_1,Movie_2),
      common_director(Movie_1,Movie_2),
      common_decade(Movie_1,Movie_2);
      two_same_plot(Movie_1,Movie_2),
      common_decade(Movie_1,Movie_2),
      one_common_spoken_language(Movie_1,Movie_2),
      black_and_white(Movie_1,Movie_2),
      one_same_actor(Movie_1,Movie_2),
      one_common_genre(Movie_1,Movie_2),
      partly_common_vote_average(Movie_1,Movie_2);
      three_same_plot(Movie_1,Movie_2),
      one_common_genre(Movie_1,Movie_2),
      common_decade(Movie_1,Movie_2),
      one_common_spoken_language(Movie_1,Movie_2).
      black_and_white(Movie_1,Movie_2),
      one_same_actor(Movie_1,Movie_2).
  ```
  ```
find_similar_7(Movie_1,Movie_2):-
      three_same_plot(Movie_1,Movie_2),
      one_common_genre(Movie_1,Movie_2),
      one_common_spoken_language(Movie_1,Movie_2),
      black_and_white(Movie_1,Movie_2),
      common_director(Movie_1,Movie_2),
      common_decade(Movie_1,Movie_2),
      one_same_actor(Movie_1,Movie_2);
      two_same_plot(Movie_1,Movie_2),
      common_decade(Movie_1,Movie_2),
      one_common_spoken_language(Movie_1,Movie_2),
      black_and_white(Movie_1,Movie_2),
      one_same_actor(Movie_1,Movie_2),
      one_common_genre(Movie_1,Movie_2),
      partly_common_vote_average(Movie_1,Movie_2),
      common_director(Movie_1,Movie_2);
      three_same_plot(Movie_1,Movie_2),
      one_common_genre(Movie_1,Movie_2),
      common_decade(Movie_1,Movie_2),
      one_common_spoken_language(Movie_1,Movie_2).
      black_and_white(Movie_1,Movie_2),
      one_same_actor(Movie_1,Movie_2),
      partly_common_budget(Movie_1,Movie_2).
  ```
  ```
find_similar_8(Movie_1,Movie_2):-
      three_same_plot(Movie_1,Movie_2),
      one_common_genre(Movie_1,Movie_2),
      one_common_spoken_language(Movie_1,Movie_2),
      black_and_white(Movie_1,Movie_2),
      common_director(Movie_1,Movie_2),
      common_decade(Movie_1,Movie_2),
      one_same_actor(Movie_1,Movie_2),
      one_common_production_studio(Movie_1,Movie_2),
      partly_common_vote_average(Movie_1,Movie_2);
      two_same_plot(Movie_1,Movie_2),
      common_decade(Movie_1,Movie_2),
      one_common_spoken_language(Movie_1,Movie_2),
      black_and_white(Movie_1,Movie_2),
      one_same_actor(Movie_1,Movie_2),
      one_common_genre(Movie_1,Movie_2),
      partly_common_vote_average(Movie_1,Movie_2),
      common_director(Movie_1,Movie_2),
      one_common_production_studio(Movie_1,Movie_2),
      partly_common_budget(Movie_1,Movie_2);
      three_same_plot(Movie_1,Movie_2),
      one_common_genre(Movie_1,Movie_2),
      common_decade(Movie_1,Movie_2),
      one_common_spoken_language(Movie_1,Movie_2),
      black_and_white(Movie_1,Movie_2),
      one_same_actor(Movie_1,Movie_2),
      partly_common_budget(Movie_1,Movie_2),
      common_director(Movie_1,Movie_2),
      one_common_production_studio(Movie_1,Movie_2).
  ```
  ```
find_similar_9(Movie_1,Movie_2):-
      three_same_plot(Movie_1,Movie_2),
      two_common_genre(Movie_1,Movie_2),
      one_common_spoken_language(Movie_1,Movie_2),
      black_and_white(Movie_1,Movie_2),
      common_director(Movie_1,Movie_2),
      common_decade(Movie_1,Movie_2),
      some_same_actors(Movie_1,Movie_2),
      one_common_production_studio(Movie_1,Movie_2),
      very_common_vote_average(Movie_1,Movie_2),
      common_duration(Movie_1,Movie_2),
      same_language(Movie_1,Movie_2),
      partly_common_budget(Movie_1,Movie_2),
      partly_common_popularity(Movie_1,Movie_2);
      two_same_plot(Movie_1,Movie_2),
      common_decade(Movie_1,Movie_2),
      one_common_spoken_language(Movie_1,Movie_2),
      black_and_white(Movie_1,Movie_2),
      some_same_actors(Movie_1,Movie_2),
      two_common_genre(Movie_1,Movie_2),
      partly_common_vote_average(Movie_1,Movie_2),
      common_director(Movie_1,Movie_2),
      one_common_production_studio(Movie_1,Movie_2),
      partly_common_budget(Movie_1,Movie_2),
      common_duration(Movie_1,Movie_2),
      same_language(Movie_1,Movie_2),
      partly_common_popularity(Movie_1,Movie_2),
      partly_common_gross(Movie_1,Movie_2);  
      three_same_plot(Movie_1,Movie_2),
      two_common_genre(Movie_1,Movie_2),
      common_decade(Movie_1,Movie_2),
      one_common_spoken_language(Movie_1,Movie_2),
      black_and_white(Movie_1,Movie_2),
      some_same_actors(Movie_1,Movie_2),
      partly_common_budget(Movie_1,Movie_2),
      common_director(Movie_1,Movie_2),
      one_common_production_studio(Movie_1,Movie_2),
      same_language(Movie_1,Movie_2),
      partly_common_popularity(Movie_1,Movie_2),
      partly_common_vote_average(Movie_1,Movie_2),
      partly_common_gross(Movie_1,Movie_2).
  ```
  ```
find_similar_10(Movie_1,Movie_2):-
      four_same_plot(Movie_1,Movie_2),
      two_common_genre(Movie_1,Movie_2),
      one_common_spoken_language(Movie_1,Movie_2),
      black_and_white(Movie_1,Movie_2),
      common_director(Movie_1,Movie_2),
      common_decade(Movie_1,Movie_2),
      some_same_actors(Movie_1,Movie_2),
      one_common_production_studio(Movie_1,Movie_2),
      very_common_vote_average(Movie_1,Movie_2),
      common_duration(Movie_1,Movie_2),
      same_language(Movie_1,Movie_2),
      very_common_budget(Movie_1,Movie_2),
      very_common_popularity(Movie_1,Movie_2),
      partly_common_gross(Movie_1,Movie_2),
      two_common_production_country(Movie_1,Movie_2),
      partly_common_num_voted(Movie_1,Movie_2).
  ```

##Ένταξη κατηγορημάτων find_similar στην Prolog

In [ ]:
prolog.consult(path + "find_similar.pl")

##Επιλεκτικές δοκιμές κατηγορημάτων find_similar

In [ ]:
q = prolog.query("find_similar_1('X-Men: First Class', Y)")
s = set()
for soln in q:
  m = soln['Y'] 
  if m not in s:
    s.add(soln['Y'])
print(s)

{'All Good Things', 'The Last Exorcism', 'Fury', 'Jimmy and Judy', 'You Dont Mess with the Zohan', 'Underdogs', 'Sphere', 'Beverly Hills Cop III', 'Limbo', 'The Sticky Fingers of Time', 'Spotlight', 'The Cave', 'Gridiron Gang', 'See Spot Run', 'Aladdin', 'H.', 'The Book of Life', 'The Ghost Writer', 'The Railway Man', 'Ultraviolet', 'From Russia with Love', 'Meet Dave', 'Muppets from Space', 'Steamboy', 'Brave', 'Penguins of Madagascar', 'Maniac', 'Unstoppable', 'Amigo', 'Instructions Not Included', 'The Switch', 'N-Secure', 'Contraband', 'Yoga Hosers', 'The Big Wedding', 'The Three Stooges', 'Enter the Dangerous Mind', 'Three Kingdoms: Resurrection of the Dragon', 'Back to the Future', 'Mad Max', 'Lara Croft Tomb Raider: The Cradle of Life', 'Avengers: Age of Ultron', 'Training Day', 'Restless', 'Stuart Little 2', 'The Wash', 'Gone Girl', 'Charlie St. Cloud', 'Robin Hood: Prince of Thieves', 'Kingdom of Heaven', 'Underworld: Rise of the Lycans', 'The Hobbit: An Unexpected Journey', 'A

In [ ]:
q = prolog.query("find_similar_4('X-Men: First Class', Y)")
s = set()
for soln in q:
  m = soln['Y'] 
  if m not in s:
    s.add(soln['Y'])
print(s)

{'Fury', 'The Equalizer', 'Spectre', 'Underdogs', 'Skyfall', 'All Superheroes Must Die', 'Swelter', 'Stolen', 'Never Let Me Go', 'All Is Lost', 'The Chronicles of Narnia: The Voyage of the Dawn Treader', 'Safe House', 'H.', 'The Perfect Wave', 'Teenage Mutant Ninja Turtles', 'The Book of Life', 'Jack Ryan: Shadow Recruit', 'Maleficent', 'Prometheus', 'Brave', 'Penguins of Madagascar', 'Dawn of the Crescent Moon', 'Unstoppable', 'Sherlock Holmes: A Game of Shadows', 'Contraband', 'The Lost Medallion: The Adventures of Billy Stone', 'Rio 2', 'Furious 7', 'Avengers: Age of Ultron', 'U.F.O.', 'Red State', 'Keanu', 'Apollo 18', 'Soul Surfer', 'Salvation Boulevard', 'Oblivion', 'The Lone Ranger', 'Big Miracle', 'The Hobbit: An Unexpected Journey', 'Gangster Squad', 'Underworld: Awakening', 'Wind Walkers', 'Noah', 'Puss in Boots', 'The Heat', 'Insurgent', '3 Days to Kill', 'Ride Along', 'Bang Bang Baby', 'Suicide Squad', 'Interstellar', 'Killer Elite', 'Teenage Mutant Ninja Turtles: Out of th

In [ ]:
q = prolog.query("find_similar_7('X-Men: First Class', Y)")
s = set()
for soln in q:
  m = soln['Y'] 
  if m not in s:
    s.add(soln['Y'])
print(s)
q.close()

{'X-Men: Days of Future Past', 'Iron Man 3', 'Iron Man 2', 'Thor: The Dark World', 'Fantastic Four', 'X-Men: Apocalypse', 'Captain America: The Winter Soldier', 'Captain America: Civil War', 'Ant-Man', 'The Avengers', 'Avengers: Age of Ultron', 'Thor', 'Man of Steel', 'Captain America: The First Avenger', 'Deadpool', 'The Wolverine'}


##Δημιουργία simple recommender με βάση τα κατηγορήματα find_similar

Με χρήση των κατηγορημάτων find_similar που κατασκευάστηκαν παραπάνω, δημιουργείται ο simple recommender, που παίρνει παίρνει ως παράμετρο το όνομα της ταινίας με βάση της οποίας θα γίνουν οι προτάσεις, ως εξης:

Αρχικοποιείται η λίστα answers με τις ταινίες που θα προταθούν, καθώς και ένα dictionary, το movies_compatability, το οποιο αντιστοιχίζει την κάθε προτεινόμενη ταινία με το επίπεδο ομοιότητάς της (αν μια ταινία Χ έχει βρεθεί πρώτα από το find_similar_8 θα ισχύει ότι movies_compatibility[Χ] = 8).

Ξεκινώντας από το μεγαλύτερο επίπεδο ομοιότητας 10 και σταδιακά μειώνοντας κατά 1 φτάνουμε μέχρι το επίπεδο 1. Για καθένα από τα επίπεδα αυτα βρίσκουμε με ένα query τις παρόμοιες με την ταινία που δώθηκε ως παράμετρος ταινίες για το αντίστοιχο επίπεδο ομοιότητας, ελέγχουμε αν έχουμε ξαναβρει την ταινία αυτή είτε στο ίδιο ειτε σε κάποιο μεγαλύτερο επίπεδο ομοιότητας. Αν την έχουμε ξαναβρεί την αγνοούμε, μιας και έχει μπει ήδη πιο ψηλά στην λίστα προτάσεων, αν πό την άλλη η ταινία δεν έχει ξαναβρεθεί την προσθέτουμε σε ένα βοηθητικό συνολο που χρησιμοποιούμε μόνο στο τρέχν επίπεδο και αποθηκεύουμε στο movies_compatibility το επίπεδο ομοιότητας της για να χρησιμοποιηθεί παρακάτω. 

Μόλις τελειώσουν όλες οι ταινίες του επιπέδου οι προτάσεις για το επίπεδο αυτό βρίσκονται στο βοηθητικό σύνολο (s10, s9, ...). Εφόσον όλες οι ταινίες του συνόλου αυτού ανήκουν στο ίδιο επίπεδο ομοιότητας δεν μας πειράζει η σχετική τους σειρά, άρα τις ταξινομούμε αλφαβητικά για να μην εμφανίζονται τυχαία και τις τοποθετούμε στην ουρά της λίστας answers που θα επιστρέψουμε στο τέλος της συνάρτησης.

Εφόσον εκτελούμε τα επίπεδα ομοιότητας ένα ένα με ανάποδη σειρά και τοποθετούμε τις προτάσεις του κάθε επιπέδου στην ουρα της λίστα answers, όλες οι προτάσεις μεγαλύτερων επιπέδων θα προηγούνται γιατί θα έχουν ήδη τοποθετηθεί στο answers σε προηγούμενο χρόνο. Με τον τρόπο αυτό διασφαλίζεται ότι στις προτάσεις που γίνονται οι ταινίες εμγανίζονται με φθίνουσα σειρά ομοιότητας.


Σημειώνουμε ότι η συνάρτηση simple_recommender που ακολουθεί τροποποιήθηκε ώστε να επιστρέφει όχι μόνο την λίστα παρόμοιων ταινιών, αλλά και ένα λεξικό που περιλαμβάνει τον βαθμό ομοιότητας (1-10) για καθεμία από τις ταινίες της λίστας.

In [ ]:
def simple_recommender(movie):
  
    movies_compatibility = {}             # dictionary shows the similarity level of a recommended movie
    answers = []


    s10 = set()                           # helper distinct set where the similar movies of level 10 that we will recommended will be temporarily stored
    q = prolog.query("find_similar_10('" + movie +"',M)")     # find similar movies (of level 10) with the movie that was passed as an argument
    for soln in q:                        # for each similar movie that was found in the above query q
        m = soln['M']                   
        if m not in s10:                  # if found the movie has not been found before 
            s10.add(soln['M'])            # add the movie to the distinct helper set
            movies_compatibility[m]=10    # also add the movie and its corresponding similarity_level (here is 10) in the movies_compatibility dictionary
    q.close()                             # close the query to ensure there will be no nested query errors 
    answers10 = list(sorted(s10))         # sort the level-10 similar movies that remained in the helper set. We can do this because they have the same similarity level
    answers.extend(answers10)             # add the sorted set to the tail of the answers, which is the list of recommended movies we want to return in the end

    s9 = set()
    q = prolog.query("find_similar_9('" + movie +"',M)")
    for soln in q:
        m = soln['M'] 
        if m not in s9 and m not in s10 : # if the movie has not been found in the same or in a higher similarity level before
            s9.add(soln['M'])             # add it to the helper set s9 that will end up being the recommended movies with similarity of level 9
            movies_compatibility[m]=9
    q.close()
    answers9 = list(sorted(s9))
    answers.extend(answers9)              # add the sorted helper set s9 to the tail of the answers, after answers of level 10 

    s8 = set()
    q = prolog.query("find_similar_8('" + movie +"',M)")
    for soln in q:
        m = soln['M'] 
        if m not in s8 and m not in s9  and m not in s10:
            s8.add(soln['M'])
            movies_compatibility[m]=8
    q.close()
    answers8 = list(sorted(s8))
    answers.extend(answers8)

    s7 = set()
    q = prolog.query("find_similar_7('" + movie +"',M)")
    for soln in q:
        m = soln['M'] 
        if m not in s7 and m not in s8  and m not in s9 and m not in s8 and m not in s10:
            s7.add(soln['M'])
            movies_compatibility[m]=7
    q.close()
    answers7 = list(sorted(s7))
    answers.extend(answers7)

    s6 = set()
    q = prolog.query("find_similar_6('" + movie +"',M)")
    for soln in q:
        m = soln['M'] 
        if m not in s6 and m not in s7 and m not in s8 and m not in s9 and m not in s10:
            s6.add(soln['M'])
            movies_compatibility[m]=6
    q.close()
    answers6 = list(sorted(s6))
    answers.extend(answers6)

    s5 = set()
    q = prolog.query("find_similar_5('" + movie +"',M)")
    for soln in q:
        m = soln['M'] 
        if m not in s5 and m not in s6 and m not in s7 and m not in s8 and m not in s9 and m not in s10:
            s5.add(soln['M'])
            movies_compatibility[m]=5
    q.close()
    answers5 = list(sorted(s5))
    answers.extend(answers5)

    s4 = set()
    q = prolog.query("find_similar_4('" + movie +"',M)")
    for soln in q:
        m = soln['M'] 
        if m not in s4 and m not in s5 and m not in s6 and m not in s7 and m not in s8 and m not in s9 and m not in s10:
            s4.add(soln['M'])
            movies_compatibility[m]=4
    q.close()
    answers4 = list(sorted(s4))
    answers.extend(answers4)

    s3 = set()
    q = prolog.query("find_similar_3('" + movie +"',M)")
    for soln in q:
        m = soln['M'] 
        if m not in s3 and m not in s4  and m not in s5 and m not in s6 and m not in s7 and m not in s8 and m not in s9 and m not in s10:
            s3.add(soln['M'])
            movies_compatibility[m]=3
    q.close()
    answers3 = list(sorted(s3))
    answers.extend(answers3)

    s2 = set()
    q = prolog.query("find_similar_2('" + movie +"',M)")
    for soln in q:
        m = soln['M'] 
        if m not in s2 and m not in s3 and m not in s4  and m not in s5 and m not in s6 and m not in s7 and m not in s8 and m not in s9 and m not in s10:
            s2.add(soln['M'])
            movies_compatibility[m]=2
    q.close()
    answers2 = list(sorted(s2))
    answers.extend(answers2)

    s1 = set()
    q = prolog.query("find_similar_1('" + movie +"',M)")
    for soln in q:
        m = soln['M'] 
        if m not in s1 and m not in s2 and m not in s3 and m not in s4  and m not in s5 and m not in s6 and m not in s7 and m not in s8 and m not in s9 and m not in s10:
            s1.add(soln['M'])
            movies_compatibility[m]=1
    q.close()
    answers1 = list(sorted(s1))
    answers.extend(answers1)

    return answers,movies_compatibility     # return the recommended movies list, as well as the movies_compatibility dictionary, which will be used in part 3

##Επιλεκτικές δοκιμές του simple_recommender

In [ ]:
#list(simple_recommender('The Avengers'))[:5]
movies,movies_comp=simple_recommender('The Avengers')
list(movies[:5])

['Avengers: Age of Ultron',
 'Ant-Man',
 'Captain America: Civil War',
 'Captain America: The First Avenger',
 'Captain America: The Winter Soldier']

In [ ]:
movies,movies_comp=simple_recommender('The Avengers')
list(movies)

['Avengers: Age of Ultron',
 'Ant-Man',
 'Captain America: Civil War',
 'Captain America: The First Avenger',
 'Captain America: The Winter Soldier',
 'Deadpool',
 'Fantastic Four',
 'Guardians of the Galaxy',
 'Iron Man 2',
 'Iron Man 3',
 'Kick-Ass 2',
 'Man of Steel',
 'The Wolverine',
 'Thor',
 'Thor: The Dark World',
 'X-Men: Apocalypse',
 'X-Men: Days of Future Past',
 'X-Men: First Class',
 'Fantastic 4: Rise of the Silver Surfer',
 'Iron Man',
 'Superman',
 'Superman II',
 'The Incredible Hulk',
 'The Shadow',
 'X-Men',
 'X-Men Origins: Wolverine',
 'X-Men: The Last Stand',
 'X2',
 '10 Cloverfield Lane',
 '127 Hours',
 '13 Hours: The Secret Soldiers of Benghazi',
 '1911',
 '2 Guns',
 '21 Jump Street',
 '22 Jump Street',
 '3 Days to Kill',
 '30 Minutes or Less',
 '300: Rise of an Empire',
 '47 Ronin',
 'A Good Day to Die Hard',
 'A Lonely Place to Die',
 'Abduction',
 'About Time',
 'Abraham Lincoln: Vampire Hunter',
 'Act of Valor',
 'After Earth',
 'Against the Wild',
 'Alex C

In [ ]:
movies,movies_comp=simple_recommender('Avatar')
list(movies)

['Aliens',
 'Alien³',
 'Galaxina',
 'John Carter',
 'Pirates of the Caribbean: On Stranger Tides',
 'Predators',
 'Star Trek Into Darkness',
 'The Fifth Element',
 '10th & Wolf',
 '12 Rounds',
 '13 Going on 30',
 '15 Minutes',
 '16 Blocks',
 '2 Fast 2 Furious',
 '2012',
 '28 Days Later',
 '28 Weeks Later',
 '3 Strikes',
 '300',
 '3000 Miles to Graceland',
 '9',
 'A Knights Tale',
 'A Man Apart',
 'A Perfect Getaway',
 'A Scanner Darkly',
 'A Sound of Thunder',
 'A.I. Artificial Intelligence',
 'AVP: Alien vs. Predator',
 'Across the Universe',
 'After the Sunset',
 'Agent Cody Banks',
 'Agora',
 'Alatriste',
 'Alexander',
 'Alien',
 'Aliens in the Attic',
 'Aliens vs Predator: Requiem',
 'All About the Benjamins',
 'All Hat',
 'All The Queens Men',
 'Alone in the Dark',
 'Along Came a Spider',
 'Alvin and the Chipmunks',
 'Alvin and the Chipmunks: The Squeakquel',
 'American Outlaws',
 'An American Carol',
 'Anacondas: The Hunt for the Blood Orchid',
 'Antitrust',
 'Apocalypto',
 'Aqua

In [ ]:
movies,movies_comp=simple_recommender('Aladdin')
list(movies)

['Pocahontas',
 'Enchanted',
 'Frozen',
 'Tangled',
 'The Illusionist',
 'The Sorcerers Apprentice',
 '10 Things I Hate About You',
 '20 Dates',
 '200 Cigarettes',
 '24 7: Twenty Four Seven',
 '3 Ninjas Kick Back',
 '30 Minutes or Less',
 '8 Heads in a Duffel Bag',
 'A Bugs Life',
 'A Dog Of Flanders',
 'A League of Their Own',
 'A Low Down Dirty Shame',
 'A Night at the Roxbury',
 'A Room for Romeo Brass',
 'A Simple Wish',
 'A Thin Line Between Love and Hate',
 'A Walk on the Moon',
 'Ace Ventura: Pet Detective',
 'Ace Ventura: When Nature Calls',
 'Air Bud',
 'Airborne',
 'Alive',
 'American Pie',
 'Among Giants',
 'An Alan Smithee Film: Burn, Hollywood, Burn',
 'An Ideal Husband',
 'Anaconda',
 'Analyze This',
 'Anastasia',
 'Anna and the King',
 'Antz',
 'Anywhere But Here',
 'Arachnophobia',
 'Armageddon',
 'Army of Darkness',
 'As Good as It Gets',
 'Assassins',
 'At First Sight',
 'Austin Powers: International Man of Mystery',
 'Austin Powers: The Spy Who Shagged Me',
 'Babe',


# **Μέρος 3: Recommendation System Με βάση τις προτιμήσεις-Αξιολογήσεις του χρήστη-Εκπαίδευση και Πρόβλεψη**






Αρχικά μελετάμε τις αξιολογήσεις κάθε χρήστη για να καταλάβουμε την δομή και τις πληροφορίες κάθε αρχείου.

Στον κώδικα που μας δόθηκε για την συνάρτηση train_recommender τροποιποίηθηκαν τα εξής:

*   Χρησιμοποιήθηκαν διαφορετικά rating_weights. Συγκεκριμένα βλέπουμε ότι τα rating weights είναι κεντρισμένα γύρω από το 3, θεωρήσαμε ότι για πολύ χαμηλές βαθμολογίες 0, 1 θα πρέπει η ταινια εν τέλει να έχει αρνητικό βάρος, για την τιμή 2 που είναι ενδιάμεση αλλα αρκετα χαμηλή βάλαμε βάρος 0, ενώ για τις τιμές από 3 σε 5 που είναι οι μεγαλύτερες βαθμολογίες έχουμε θετικά βάρη ξεκινώντας για το 3 με τιμή κοντά στο 0 και ανεβαίνοντας όλο και περισσότερο κάθε φορα για να φτάσουμε στο 5 .

*   Τροποποίηθηκαν επίσης τα βάρη score_weights. Σύμφωνα με τα επίπεδα ομοιότητας που έχουμε ορίσει θεωρούμε ότι κάθε επίπεδο έχει διαφορετικό βάρος και πως σε μεγαλύτερα επίπεδα έχει μεγαλύτερη αξία να περάσει μια ταινία σε μεγαλύτερο επίπεδο ομοιότητας. Δηλαδή θεωρήθηκε ότι αν μια ταινία περάσει από το επίπεδο 7 στο επίπεδο 8 είναι πολύ πιο όμοια με την ζητούμενη, άρα θα πρέπει να κερδίσει περισσότερο βάρος από ότι όταν θα περάσει από το επίπεδο 1 στο επίπεδο 2. Για αυτό τον λόγο τα βάρη δεν έχουν οριστεί με σταθερό βήμα. Συγκεκριμένα :
      *    Για επίπεδα ομοιότητας από 1 μέχρι 4 οι ταινίες που επιστρέφονται είναι σχετικά γενικές. Οπότε δίνουμε μια πολύ μικρή τιμή βάρους 0,5 για τις πολύ γενικές ταινίες επιπέδου 1 και όσο ανεβαίνουμε επίπεδο οι ταινίες γίνονται ελαφρώς πιο συγκεκριμένες, άρα βάζουμε ένα μικρό βήμα στην αύξηση των βαρών, αξίας 0,5 .
      *    Φτάνοντας στα επίπεδα 5 και 6 προστιθενται αρκετοί κανόνες, άρα οι ταινίες έχουν αρχίσει να γίνονται πιο συγκεκριμένες, για αυτό και το βήμα αύξησης βαρών τώρα γίνεται μεγαλύετερο και ίσο με 1.
      *    Τέλος για τα επίπεδα 7, 8 και 9 το σύνολο των κανόνων έχει γίνει αρκετά μεγάλο, οπότε οι ταινίες είναι πλέον πολύ όμοιες με την ζητούμενη και επιπλέον είναι πολύ πιο δύσκολο μια ταινία να φτάσει μέχρι τα επίπεδα αυτά και να περάσει από το ένα επίπεδο στο άλλο. Για αυτό και τα βάρη τους έχουν μεγαλύτερες τιμές, αλλά και η αξία του βήματος μεταξύ των επιπέδων αυτών έχει μεγαλύτερη αξία, ίση με 1,5 .
      *     Για το επίπεδο 10 όλα τα if-else clauses ενωποιούνται και πρέπει όλοι οι κανόνες να ικανοποιούνται ταυτόχρονα. Επομένως για τον λόγο αυτό θεωρούμε ότι το επίπεδο ομοιότητας αυτό έχει πολύ μεγαύτερο βάρος από όλα τα προηγούμενα του και για αυτό παιρνει την μεγαλύτερη τιμή με απόσταση από το προηγούμενο επίπεδο ομοιότητας 9 ίση με 2,5 .

*  Η τελευταία τροποποίηση που έγινε είναι πως για να βρούμε το επίπεδο ομοιότητας της κάθε ταινίας χρησιμοποιείται το movies_compatibility που ορίστηκε στην παραπάνω συναρτηση simple recommender, το οποίο εδώ έχει όνομα μεταβλητής similarity_dictionary. Επομένως για να βρεθεί το βάρος που αντιστοιχεί κάθε φορά στο συγκεκριμένο επίπεδο ομοιότητας εκτελείται το :  score_weights[ similarity_dict[similar_movie] ] .



In [ ]:
from tqdm.notebook import tqdm
from sklearn.metrics import precision_score, recall_score, f1_score
import numpy as np
import random

rating_weights = {0: -1, 1: -0.5, 2:0, 3:0.4, 4:0.5, 5:1}
score_weights = {1: 0.5, 2: 1, 3: 1.5, 4: 2, 5: 3, 6: 4, 7: 5.5, 8: 7, 9: 8.5, 10: 11} 

#score_weights = {i:i + 1 for i in range(10)} # ανάλογα με τα επίπεδα ομοιότητας που έχουν οριστεί στην simple_recommender
#rating_weights = {0: -1.5, 1: -1, 2:-0.5, 3:0.5, 4:1, 5:1.5}

def train_recommender(ratings, rating_weights, score_weights, number_of_movies = 10):
    """
    Στην συνάρτηση αυτή μπορούμε να ορίζουμε ποιο υποσύνολο των αξιολογήσεων θα χρησιμοποιήσουμε για το train μαζί με τα βάρη ομοιότητας και σκορ
    Σε συνδυασμό με τον αριθμό των ταινιών που θέλουμε να χρησιμοποιήσουμε σαν σύνολο δεδομένων π.χ. 10 από τις 100 ή 3 από τις 100 κ.ο.κ
    Αν θέλουμε να χρησιμοποιήσουμε όλες τις ταινίες σαν training set τότε ορίζουμε το number_of_movies = - 1
    """

    if number_of_movies > len(ratings):
        number_of_movies = len(ratings)


    if number_of_movies != -1:
        indexes = random.sample(range(len(ratings)), number_of_movies)
        ratings = ratings.iloc[indexes]

    movie_score = {}
    for row in tqdm(ratings.itertuples(index=True, name='Pandas')):
        movie = clean_text(getattr(row, 'movie_title'))
        rating = getattr(row, 'rating')

        similar_movies, similarity_dict = simple_recommender(movie)

        for similar_movie in similar_movies:
            if similar_movie not in movie_score:
                movie_score[similar_movie] = rating_weights[int(rating)] * score_weights[similarity_dict[similar_movie]]
            else:
                movie_score[similar_movie] += rating_weights[int(rating)] * score_weights[similarity_dict[similar_movie]] #το weight θα το ορίσετε ανα επίπεδο ομοιότητας οι πολύ όμοιες ταινίες θα έχουν μεγαλύτερο βάρος
    return movie_score


#αυτό είναι ένα παράδειγμα για το πως θα μπορούσε να υλοποιήθει η predict
#έχουμε ορίσει ότι μια ταινία θα έπρεπε να είναι προτεινόμενη αν είχε σκορ > 0

def predict_example(ratings, movie_score):
    real, pred = [], []
    for i, row in enumerate(ratings.itertuples(index=True, name='Pandas')):
        movie = clean_text(getattr(row, 'movie_title'))
        rating = getattr(row, 'rating')

        if movie in movie_score:                        #αν έχουμε σχηματίσει βαθμολογία για την ταινία αυτή
            pred.append(int(movie_score[movie] > 0))    #heuristic για το αν μια ταινία είναι προτεινόμενη
            real.append(int(rating > 3))                # έτσι ορίζουμε ότι μια ταινία θα έπρεπε να είναι προτεινόμενη
            #η συνθήκη αυτή δεν μπορεί να αλλάξει
        else:                                           #δεν μπορούμε να προτείνουμε κάτι για το όποιο δεν έχουμε σχηματίσει εικόνα
            pred.append(0)
            real.append(int(rating > 3))

    return real, pred


def get_metrics(real, pred):
    metrics = {}
    metrics["precision"] = precision_score(real, pred)
    metrics["recall"] = recall_score(real, pred)
    metrics["f1"] = f1_score(real, pred)
    return metrics

Η παραπάνω συναρτήσεις εκπαιδεύουν, τεστάρουν και μετρούν την απόδοσή του συστήματος συστάσεων μας. Για την εκπαίδευση του συστήματος μπορούμε κάθε φορά να χρησιμοποιήσουμε ένα τυχαίο υποσύνολο του training set. Όμως είναι πιθανό το υποσύνολο των ταινιών αυτών να επηρεάζει τα αποτελέσματα στο training set. Για παράδειγμα από 3 ταινίες στις 10 μπορούν για ένα πείραμα τα αποτελέσματα μας να είναι ίδια και αυτό να μην οφείλεται στο γεγονός ότι ο ταξινομητής μας τα πηγαίνει καλά στις 3 ταινίες αλλά στο γεγονός ότι οι υπόλοιπες 7 είναι τέτοιες ταινίες που δεν μας βοηθούν καθόλου. Αν αντίστοιχα είχαμε επιλέξει διαφορετικές 3 ταινίες θα μπορούσαμε να τα έχουμε πάει πολύ χειρότερα. Επομένως, τρέχουμε κάθε φορά έναν αριθμό πειραμάτων για κάθε υποσύνολο ταινιών, συγκεκριμένα 10 πειράματα για κάθε επιλεγμένο αριθμό ταινιών, και κρατάμε σαν τελικό σκορ το μέσο όρο όλων των πειραμάτων. Έτσι τα αποτελέσματα είναι πιο αντικειμενικά.

Για την εκπαίδευση και αξιολόγηση του συστήματος συστάσεών μας φορτώνουμε τα απαραίτητα αρχεία:

In [ ]:
train_ratings = pd.read_csv(path + "train_ratings.csv")
test_ratings = pd.read_csv(path + "test_ratings.csv")

Ακουλουθεί η εκπαίδευση και η αξιολόγηση του συστήματος συστάσεων, χρησιμοποιώντας με χρήση διαφορετικού πλήθους ταινιών για την εκπαίδευσή του κάθε φορά:

In [ ]:
# using 3 movies:
metrics = []
for i in range (10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 3)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.4925971657915036
recall: 0.7375
f1: 0.5739336208741748


In [ ]:
# using 5 movies:
metrics = []
for i in range (10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 5)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.5043211259814393
recall: 0.9027777777777779
f1: 0.645114411206875


In [ ]:
# using 7 movies:
metrics = []
for i in range (10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 7)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.5136858857125526
recall: 0.898611111111111
f1: 0.6510883711598908


In [ ]:
# using 10 movies:
metrics = []
for i in range (10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 10)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.5098717496149839
recall: 0.9361111111111111
f1: 0.6592948913203677


In [ ]:
# using 20 movies:
metrics = []
for i in range (10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 20)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.5093126428590723
recall: 0.9888888888888889
f1: 0.6723288322925536


In [ ]:
# using 30 movies:
metrics = []
for i in range (10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 30)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.5096371688396847
recall: 0.9902777777777778
f1: 0.6729159434958281


In [ ]:
# using 50 movies:
metrics = []
for i in range (10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 50)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.5099392097264437
recall: 0.9972222222222221
f1: 0.6748073345734786


## Σχολιασμός και Παρατηρήσεις

Όπως φαίνεται παραπάνω καλούμε την συνάρτηση train recommender για υποσύνολα ταινιών διαφορετικών μεγεθών και συγκεκριμένα διαδοχικά για υποσύνολα μεγέθους 3, 5, 7, 10, 20, 30 και 50 ταινίες. Σημειώνεται ότι για καθένα από τα παραπάνω μεγέθη εκτελούμε 10 πειράματα και στο τέλος παρουσιάζεται ο μέσος όρος των πειραμάτων αυτών, ώστε να έχουμε πιο αντικειμενικά αποτελέσματα.

Παρατηρείται από τα παραπάνω αποτελεσματα ότι η τιμή του recall είναι πολύ υψηλή ακόμα και σε πειράματα με λίγες ταινίες ως σύνολο εκπαίδευσης. Αυτό σημαίνει πως ο Recommender μας είναι σε θέση να προτείνει πολλές από τις ταινίες που θα έπρεπε να προτείνει, ακόμα και με μικρό train set, γεγονός που προφανώς οφείλεται στον τρόπο με τον οποίο έχουμε ορίσει τα find similar κατηγορήματα, τα οποία φαίνεται όντως να εντοπίζουν όλες τις ταινίες που πρέπει. 

Ωστόσο παρατηρείται ότι η τιμή του precision είναι για όλα τα υποσύνολα εκπαίδευσης της τάξης του 50% , κάτι το οποίο σημαίνει ότι ναι μεν ο recommender μας κάνει όλες τις απαραίτητες προτάσεις αλλα φαίνεται να κάνει πολύ περισσότερες από αυτές. Αναλυτικότερα η παραπάνω τιμή του precision σημαίνει ότι από τις ταινίες που προτείναμε επρέπε να προταθεί μόνο το 50% . Η τιμή του precision δεν είναι ιδανική αλλά είναι σίγουρα αποδεκτή. 

Η τιμές του F1 score, ως μία εξισορρόπηση των δύο παραπάνω μετρικών, βρίσκεται ανάμεσα στις τιμές αυτών και ξεκινά από 57% περίπου για μικρό σύνολο εκπαίδευσης, φτάνοντας για περισσότερες ταινίες το 67% περίπου. Πρόκειται για μία τιμή η οποία σίγουρα δεν είναι βέλτιστη, δείχνει όμως πως το σύστημα έχει αρκετά καλές επιδόσεις.

Αναλύοντας περαιτέρω τα αποτελέσματα των τριών μετρικών, συμπεραίνουμε ότι ο recommender μας είνα αρκετά "ανεκτικός" στο ποιες ταινίες θεωρεί όμοιες. Αυτό έχει ως αποτέλεσμα όλες οι όμοιες ταινίες να βρίσκονται σωστά (υψηλό recall), αλλά και περίπου άλλες τότσες ταινίες, οι οποίες ίσως δεν μοιάζουν τόσο, να θεωρούνται από τον recommender παρόμοιες (χαμηλότερο precision).

Θα μπορούσαμε ίσως να βελτιώσουμε το precision κάνοντας τους κανόνες find similar πιο αυστηρούς, ώστε να βρίσκουν λιγότερες και περισσότερο όμοιες ταινίες, ή θα μπορούσαμε να τροποποιήσουμε τα score_weights και rating_weights, ώστε οι ταινίες χαμηλότερου επιπέδου ομοιότητας να έχουν λιγότερο βάρος ή οι χαμηλές βαθμολογίες να μετράνε περισσότερο αρνητικά. Όλες οι επιλογές που αναφέρθηκαν θα είχαν ως αποτέλεσμα οι προτεινόμενες ταινίες να γίνουν λιγότερες. Κάτι τέτοιο θα μπορούσε να επιδράσει θετικά στην μετρική presicion, καθώς είναι πιθανό πολλές από τις ταινίες που θα απορρίπτονταν πλέον από το σύνολο να ήταν ταινίες που δεν έπρεπε εξ αρχής να προταθούν. Από την άλλη όμως, δεν είναι σίγουρο πως το αποτέλεσμα στο presicion θα ήταν θετικό: Αν ανάμεσα στις ταινίες που απορρίπτονταν βρίσκονταν ταινίες οι οποίοι θα έπρεπε να είχαν προταθεί, τότε η τιμή του presicion θα έπεφτε. Επιπλέον, στην περίπτωση αυτή πιθανόν να μειωνόταν η τιμή του recall αφού, όπως αναφέραμε, είναι πιθανό κάποιες από τις ταινίες που απορρίφθηκαν να έπρεπε εν τέλει να προταθούν.
Πράγματι, μετά από σχετικές δοκιμές, διαπιστώσαμε ότι εφαρμόζοντας τις παραπάνω ιδέες δεν υπήρξε σημαντική βελτίωση του precision, ενώ το recall μειώθηκε αισθητά. Αποφασίστηκε λοιπόν, εφόσον η τιμή του recall είναι εξαιρετικά υψηλή και η τιμή του precision είναι αποδεκτά καλή με την παρούσα υλοποίηση, να μην γίνει κάποια από τις προαναφερθείσες τροποποιήσεις. Αξίζει να σχολιάσουμε ότι ένας παράγοντας που μπορεί εμποδίζει την βελτίωση της συνολικής επίδοσης του συστήματός μας καθώς το κάνουμε πιο αυστηρό είναι η φύση αυτού: Πρόκειται για ένα σύστημα που θέτει ορισμένους κανόνες ομοιότητας οι οποίες είναι πολύ απόλυτοι και όμοιοι για όλους τους χρήστες. Όμως, καθώς κάθε χρήστης είναι διαφορετικός, μπορεί να είναι διαφορετικά και τα κριτήριά του: Ένας χρήστης θα μπορούσε να είναι fan ενός συγκεκριμένου ηθοποιού ή σκηνοθέτη και να αξιολογεί θετικά όλες τις ταινίες αυτού, ενώ ένας άλλος μπορεί να μην ενδιαφέρετια για κάτι τέτοιο, να έχει όμως μια σαφή προτίμηση σε ταινίες εποχής. Έτσι, οι προτάσεις που αναμένει να δει κάθε χρήστης διαφέρουν. Κάνοντας τον recommender πιο "ανεκτικό", μπορούμε να κάνουμε προτάσεις που ταιριάζουν σε πολλά διαφορετικά προφίλ χρηστών. Αυτό σημαίνει πως θα προτείνουμε πολλές ταινίες που θα αρέσουν στον κάθε χρήστη, αναπόφευκτα όμως θα προτείνουμε και άλλες που δεν του αρέσουν. Αντίθετα, κάνοντας το σύστημα πιο συγκεκριμένο, μπορεί για κάποιον χρήστη αυτό να είναι θετικό, και να περιορίσουμε τις προτάσεις που δεν του ταιριάζουν, είναι όμως πολύ εύκολο να παραβλέψουμε κάποια κριτήρια που θέτει ένας χρήστης για το ποια ταινία θα του αρέσει, και να προτείνουμε λιγότερες ή λάθος ταινίες σε αυτόν.

Παρατηρείται ότι η τιμή του recall μεγιστοποιείται στο 99% ενώ του precision στο 51% , ενώ η συνδυαστική τιμή f1 μεγιστοποιείται στην τιμή 67% . Παρατηρούμε επίσης ότι οι τιμές αυτές προσεγγίζονται πολύ γρήγορα, για μικρά μεγέθη υποσυνόλων εκπαίδευσης. Ήδη για 10 ταινίες οι τιμή του recall έχει φτάσει το 93% και του precision το 50% ενώ η συνδυαστική τους τιμή f1 ανέρχεται στο 65% , για μέγεθος υποσυνόλου εκπαίδευσης 20 ταινιών το recall έχει ανέβει στο 98% , το Precision είναι της τάξης του 50% και η συνδυαστική τους τιμή f1 έχει φτάσει ήδη στο 67% . Για υποσύνολα μεγέθους μεγαλύτερου των 20 ταινιών η τιμή των recall, precision και η συνδυαστική τιμή f1 φαίνεται να σταθεροποιούται. 

Το γεγονός ότι φτάνουμε σχεδόν στις μέγιστες τιμές recall, precision και f1 για τόσο μικρό μέγεθος υποσυνόλου επαίδευσης (20 ταινίες ή ακόμα και οι 10 ταινίες) μας οδηγεί στο συμπρέρασμα ότι, ο recommender που εν τέλει δημιουργήσαμε, είναι ιδιαίτερα κατάλληλος για νέους χρήστες. Οι νέοι χρήστες δεν έχουν βαθμολογήσει πολλές ταινίες, επομένως δεν έχουμε πολλά δεδομένα για αυτούς, ωστόσο παρατηρείται ότι ο recommender μας ακόμα και για πολύ μικρά training set, δηλαδή πολύ λίγα δεδομένα όσον αφορά τις προτιμήσεις του χρήστη, έχει καλή απόδοση και προτείνει σχεδόν όλες τις ταινίες που πρέπει. Προτείνει βέβαια και πολλές επιπλέον, κάτι τέτοιο θεωρούμε όμως αρκετά αποδεκτό για έναν νέο χρήση. Όσον αφορά παλαιότερους χρήστες, ο recommender εξακολουθεί να έχει αρκετά ικανονποιητική επίδοση (με F1 score 67% περίπου), δεν βελτιώνει όμως την ακρίβειά του σηματνικά σε σχέση με όταν ο χρήστης ήταν πιο νέος, γεγονός που σίγουρα δεν είναι βέλτιστο, καθώς ένας χρήστης μπορεί να είναι πιο δεκτικός σε λάθη του συστήματος στην αρχή, αναμένει όμως συνήθως να δει αισθητή βελτίωση στις προτάσεις που λαμβάνει, όσο μεγαλύτερη ανατροφοδότηση παρέχει στο σύστημα.

